library(shiny)
library(ggplot2)
library(Seurat)
library(shinythemes)
library(TissueEnrich)
library(SummarizedExperiment)
library(igraph)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(org.Hs.eg.db)
library(karyoploteR)
library(shinyalert)
library(dorothea)
library(textshape)
library(tidyr)
library(tidyverse)
library(reshape2)
library(data.table)
library(networkD3)
library("ggplot2")
library("shinyFiles")
library(dorothea)
library(bcellViper)
library(dplyr)
library(viper)
library(cli)
options(shiny.maxRequestSize = 10000*1024^2) 
feta_name=c("Ectodermal cell differentiation","Ectodermal development","Embryonic_stem_cell","Adult b_cell","Adult b_cell_plasmocyte","Fetal fibroblast","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult adrenal gland inflammatory cell","Adult monocyte","Adult AT2 cell","Adult basal_cell","Adult CB CD34+23","Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
abc=list()
for( i in feta_name)
{ 
  fetal_name=read.csv(paste("./",i,".tsv", sep =""))
  fetal_name<-unique(fetal_name[,1])
  fetal_list=as.vector(fetal_name)
  
  abc[[i]]=GeneSet(fetal_list,setName=i)
}
gensets<-GeneSetCollection(abc)

shinyServer(function(input,output){
  
  
  
  
  
  
  observeEvent(input$testme,{
    if (input$testme) {
      hide('user_input')
    } else if (input$testme)
    {
      hide('user_input')
      hide("testme")
    }
    else {
      show('user_input')
    }
  })
  data <- reactive({
    if (input$testme) {
      # test file
      inputfile <- read.csv("gse_new.csv", header= TRUE, sep=',',row.names = as.numeric(TRUE),stringsAsFactors = FALSE)
    } 
    
 #  else if(input$testme2) 
   #  {
    # inputfile <- read.csv("gsm.csv", header= input$header, sep=',',row.names = as.numeric(input$rowname))
     
   #}
    else if(! is.null(input$File10)){
      req(input$File10)
      File10 <- input$File10
      filesdir = dirname(File10[1,4])
      
      #inFile = inFile$datapath
      
      file.rename(File10$datapath[1],paste0(filesdir,'/',File10$name[1]))
      file.rename(File10$datapath[2],paste0(filesdir,'/',File10$name[2]))
      file.rename(File10$datapath[3],paste0(filesdir,'/',File10$name[3]))
      
      pbmc.data <- Read10X(data.dir = filesdir)
    }
    
    else {
      # user inputs count matrix
      req(input$file)
      file1 <- input$file
      #if(is.null(file1)){return()} 
      inputfile<- read.table(file=file1$datapath, header= TRUE, sep=',',row.names = as.numeric(TRUE),stringsAsFactors = FALSE)}
    
  })

data1<-reactive({
  if (input$testme3) {
    # test file
    inputfile <- read.csv("meta.csv", header= TRUE, sep=',',row.names = as.numeric(TRUE),stringsAsFactors = FALSE)
  } 
 else { 
  req(input$meta)
  
  meta1<-input$meta
inputfile<- read.csv(file=meta1$datapath,sep=",")}
 
   })

output$merge_meta<-renderPrint({
  if(input$m1==0){
    return()
  }
  cluster_ident<-pb@active.ident
  
  cluster_ident<-as.data.frame(cluster_ident)
  #names(cluster_ident)[1] <- "Sample"
  write.csv(cluster_ident,file="cluster_ident.csv",row.names = TRUE)
  
  cluster_ident$Sample<-row.names(cluster_ident)
  cluster_ident
  meta_m<-data1()
 # seurat_m<-as.data.frame(seurat_m)
  #seurat_m<-t(seurat_m)
  #seurat_m<-as.data.frame(seurat_m)
 # seurat_m<-setDT(seurat_m,keep.rownames="...1")
  merger_m<-merge(cluster_ident,meta_m,by.x="Sample",by.y="Sample")
  merger_m
})

meta_file<-reactive({
  cluster_ident<-pb@active.ident
  
  cluster_ident<-as.data.frame(cluster_ident)
  #names(cluster_ident)[1] <- "Sample"
  write.csv(cluster_ident,file="cluster_ident.csv",row.names = TRUE)
  
  cluster_ident$Sample<-row.names(cluster_ident)
  cluster_ident
  meta_m<-data1()
  # seurat_m<-as.data.frame(seurat_m)
  #seurat_m<-t(seurat_m)
  #seurat_m<-as.data.frame(seurat_m)
  # seurat_m<-setDT(seurat_m,keep.rownames="...1")
  merger_m<-merge(cluster_ident,meta_m,by.x="Sample",by.y="Sample")
  merger_m
}) 



  
output$dirp<-renderPrint({
  if(input$A2==0){
    return()
  }
  req(input$File10)
  File10 <- input$File10
  filesdir = dirname(File10[1,4])
  
  #inFile = inFile$datapath
  
  file.rename(File10$datapath[1],paste0(filesdir,'/',File10$name[1]))
  file.rename(File10$datapath[2],paste0(filesdir,'/',File10$name[2]))
  file.rename(File10$datapath[3],paste0(filesdir,'/',File10$name[3]))
  
  pbmc.data <- Read10X(data.dir = filesdir)
  pb <<- CreateSeuratObject(counts = pbmc.data, min.cells = 3, min.features = 200,project = "scRNAseq")
  print("Your object has been created, start downstream analysis now!!!!")
  
  
})
   
 
  output$dat<-renderTable({
    if(is.null(data())){return ()}
    data()[1:10,1:7]
  })
  
  output$dimen <- renderPrint({
    print("Dimensions")
    dim(data())
  })
  observeEvent(input$action1, {
    # Show a modal when the button is pressed
    
    shinyalert(animation = TRUE,
               imageUrl = "aayu_popup.gif",
               imageWidth = 400,
               imageHeight = 350,confirmButtonText="OK",confirmButtonCol="black")
    #src = "gig.gif"
  })
  
  observeEvent(input$action11, {
    # Show a modal when the button is pressed
    shinyalert(title = " Markers are being generated... Please check Markers table", type = "success")
  })
  
 # observeEvent(input$action1001, {
    # Show a modal when the button is pressed
  #  shinyalert(title = " Object created. Go to scAnalysis tab for further analysis....", type = "success")
 # })
  
  output$seu_object <- renderPrint({
    if(input$action1==0){
      return()
    }
    
    pb <<- CreateSeuratObject(counts = data(), min.cells = 3, min.features = 200,project = "scRNAseq")
    print(pb)
    print("Your object has been created, start downstream analysis now!!!!")
  })
 
    
  output$ex_object<-renderPrint({
    if(input$action1==0){
      return()
    } 
    paste("The object serves as a container that contains both data (like the count matrix) and analysis (like PCA, or clustering results) for a single-cell dataset.")
  })
  
  vln<-reactive({
    VlnPlot(pb, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
    
  })
  output$p1<-renderPlot({
    if(input$action1==0){
      return()
    }
      
    
   VlnPlot(pb, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
    
  })
  output$ex_variable<-renderPrint({
    if(input$action1==0){
      return()
    } 
    paste("It is to visualize QC metrics as a violin plot. It visualize QC metrics, and use these to filter cells.")
  })
  
  output$dndPlot <- downloadHandler(
    filename = function(){'VLN_plot.pdf'},
    content = function(file){
      pdf(file)
      print(    vln()
      )
      dev.off()
    })
  
  
  observeEvent(input$toTop, {
    js$toTop();
  })
  observeEvent(input$toTop1, {
    js$toTop();
  })
  observeEvent(input$toTop3, {
    js$toTop();
  })
  observeEvent(input$toTop4, {
    js$toTop();
  })
  observeEvent(input$toTop5, {
    js$toTop();
  })
  observeEvent(input$toTop2, {
    js$toTop();
  })
  observeEvent(input$toTop6, {
    js$toTop();
  })
  observeEvent(input$toTop7, {
    js$toTop();
  })
  observeEvent(input$toTop8, {
    js$toTop();
  })
  observeEvent(input$toTop100010, {
    js$toTop();
  })
  output$data_norm<-renderPrint({
    if(input$action1==0){
      return()
    }
    print("Normalization Done")
  })
  output$ex_normalization<-renderPrint({
    if(input$action1==0){
      return()
    } 
    paste(" It employs a global-scaling normalization method LogNormalize that normalizes the feature expression measurements for each cell by the total expression, multiplies this by a scale factor, and log-transforms the result.")
  })
  
  output$variable_features<-renderPrint({
      pb <<- FindVariableFeatures(pb, selection.method = "vst", nfeatures = 2000)
      #list_of_variable_features<<-VariableFeatures(pb)
      #list_of_variable_features<<-as.data.frame(list_of_variable_features)
      top10 <<- head(VariableFeatures(pb), 20)
  })
  

  output$ex_vfeature<-renderPrint({
    if(input$action1==0){
      return()
    } 
    print("Variable feature helps in subsetting of features that exhibit high cell-to-cell variation in the dataset (i.e, they are highly expressed in some cells, and lowly expressed in others).These genes in downstream analysis helps to highlight biological signal in single-cell datasets.")
  })
  
  
  variableplot<-reactive({
    pb <<- NormalizeData(pb)
    
    pb <<- FindVariableFeatures(pb, selection.method = "vst", nfeatures = 2000)
    
    top10 <<- head(VariableFeatures(pb), 20)
    plot1 <- VariableFeaturePlot(pb)
    plot2 <- LabelPoints(plot = plot1, points = top10,repel=TRUE,xnudge = 0,ynudge = 0)
    plot(plot2) 
  })
  
  
  
  
  output$plot_vf<-renderPlot({
    if(input$action1==0){
      return()
    }
    pb <<- NormalizeData(pb)
    
    pb <<- FindVariableFeatures(pb, selection.method = "vst", nfeatures = 2000)
    top10 <<- head(VariableFeatures(pb), 20)
      plot1 <- VariableFeaturePlot(pb)
      plot2 <- LabelPoints(plot = plot1, points = top10,repel=TRUE,xnudge = 0,ynudge = 0)
     plot(plot2)
  })
  
  
  output$vfPlot <- downloadHandler(
    filename = function(){'Vf_plot.pdf'},
    content = function(file){
      pdf(file)
      print(       variableplot()
      )
      dev.off()
    })
  

    
  output$scale_data<-renderPrint({
    if(input$action1==0){
      return()
    }
   
    print("Centering and scaling data matrix")
    print("100% Done")
  })
  output$ex_scale<-renderPrint({
    if(input$action1==0){
      return()
    } 
    print("It is a linear transformation ('scaling') that is a standard pre-processing step prior to dimensional reduction techniques like PCA. It shifts the expression of each gene so that the mean expression across cells is 0 and scales the expression of each gene,  so that the variance across cells is 1. This step gives equal weight in downstream analyses so that highly-expressed genes do not dominate.")
  })
  
    
    output$pca<-renderPlot({
      if(input$action1==0){
        return()
      }
      all.genes <- rownames(pb)
      pb <<- ScaleData(pb, features = all.genes)
      pb <<- RunPCA(pb, features = VariableFeatures(object = pb),npcs = 48)
      DimPlot(pb, reduction = "pca", pt.size = 4, group.by = "ident")
    })
  pcaplot<-reactive({
    
    all.genes <- rownames(pb)
    pb <<- ScaleData(pb, features = all.genes)
    pb <<- RunPCA(pb, features = VariableFeatures(object = pb),npcs = 48)
    DimPlot(pb, reduction = "pca", pt.size = 4, group.by = "ident") 
  })  
    
 
  output$pcaPlot <- downloadHandler(
      filename = function(){'PCA_plot.pdf'},
      content = function(file){
        pdf(file)
        print(        pcaplot()
                    
        )
        dev.off()
      })
    
  output$ex_pca<-renderPrint({
    if(input$action1==0){
      return()
    }  
    paste("It performs a linear dimensional reduction")
  })
    
 heatmapplot<-reactive({
       DimHeatmap(pb, dims = 1:6, cells = 500, balanced = TRUE)
  
 }) 
  
  
    output$heatmap<-renderPlot({
      if(input$action1==0){
        return()
      }
    
      DimHeatmap(pb, dims = 1:6, cells = 500, balanced = TRUE)
      
    })
    output$ex_heatmap<-renderPrint({
      if(input$action1==0){
        return()
      }
      paste("It allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and features are ordered according to their PCA scores. 
")
    })
    output$ex_karyo<-renderPrint({
      if(input$action1==0){
        return()
      }
 paste("It gives user the chromosomal location of the selected gene.")
    })
    output$heatmapPlot <- downloadHandler(
      filename = function(){'HEATMAP_plot.pdf'},
      content = function(file){
        pdf(file)
        print(         
          heatmapplot())
        dev.off()
      })
    
    jack<-reactive({
      pb <<- JackStraw(pb, num.replicate = 100)
      pb <<- ScoreJackStraw(pb, dims = 1:20)
      JackStrawPlot(pb, dims = 1:15)
    })
    output$jackstraw<-renderPlot({
      if(input$action1==0){
        return()
      }
      pb <<- JackStraw(pb, num.replicate = 100)
      pb <<- ScoreJackStraw(pb, dims = 1:20)
      #pdf(file="Main_pipeline/GSE756881/Jackstraw_plot.pdf")
      JackStrawPlot(pb, dims = 1:15)
    })
    output$ex_jackstraw<-renderPrint({
      if(input$action1==0){
        return()
      }  
       paste("The JackStrawPlot function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). 'Significant' PCs will show strong enrichment of features with low p-values (solid curve above the dashed line")
          })
    
    output$jackstrawPlot <- downloadHandler(
      filename = function(){'Jackstraw_plot.pdf'},
      content = function(file){
        pdf(file)
        print(jack()

          )       
        dev.off()
      })
  
    output$ex_elbow<-renderPrint({
      if(input$action1==0){
        return()
      }  
      paste("A ranking of principle components based on the percentage of variance explained by each one")
    })
    output$elbow<-renderPlot({
      if(input$action1==0){
        return()
      }
      ElbowPlot(pb)
      
    })
    
    
    elbow<-reactive({
      ElbowPlot(pb)
    })
    output$elbowPlot <- downloadHandler(
      filename = function(){'elbow_plot.pdf'},
      content = function(file){
        pdf(file)
        print(       elbow()
                   
        )       
        dev.off()
      })
    
    
    output$neigh<-renderPrint({
      if(input$action1==0){
        return()
      }
      pb <<- FindNeighbors(pb, dims = 1:10)
      pb <<- FindClusters(pb, resolution = 0.5)
      
       head(Idents(pb), 5)
      
    })
    output$ex_umap<-renderPrint({
      if(input$action1==0){
        return()
      }  
paste("The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots.")
          })
    output$ex_umapp<-renderPrint({
      if(input$action1==0){
        return()
      }  
      paste("The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots.")
    
      })
    output$ex_stouffer<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Box plot depicting Stouffer's score computed based on an interaction between related genes across the indicated cell types.")
    })
    
    
    output$ex_stouffer_hpa<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Box plot depicting Stouffer's score computed based on an interaction between related genes across the indicated cell types.")
    }) 
    output$ex_stouffer_gtex<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Box plot depicting Stouffer's score computed based on an interaction between related genes across the indicated cell types.")
    })
    output$ex_karyo<-renderPrint({
      if(input$action1==0){
        return()
      } 
     paste("It is to determine the position of some genes in the genome (i.e. chromosomal location)")
          })
    output$ex_hpagrn<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Gene regulatory method helps user to understand with the help of which transcrition factor the target gene is expressing in which particular signature type. Two graphs are plotted which shows the interaction between transcription factors and target genes (Green links) and other between target genes and signatures (pink links)")
    })
    output$ex_hpagrn_table<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("The table represents the target genes with their respective transcription factor and signature. Here mor is mode of regulation and -1,+1 represents the modes and confidence level from A-C where A means higher confidence.")
    })
    output$ex_gtexgrn<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Gene regulatory method helps user to understand with the help of which transcrition factor the target gene is expressing in which particular signature type. Two graphs are plotted which shows the interaction between transcription factors and target genes (Green links) and other between target genes and signatures (pink links)")
    })
    output$ex_gtexgrn_table<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("The table represents the target genes with their respective transcription factor and signature. Here mor is mode of regulation and -1,+1 represents the modes and confidence level from A-C where A means higher confidence.")
    })
    output$ex_cellgrn<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("Gene regulatory method helps user to understand with the help of which transcrition factor the target gene is expressing in which particular signature type. Two graphs are plotted which shows the interaction between transcription factors and target genes (Green links) and other between target genes and signatures (pink links)")
    })
    output$ex_cellgrn_table<-renderPrint({
      if(input$action1==0){
        return()
      } 
      paste("the table represents the target genes with their respective transcription factor and signature. Here mor is mode of regulation and -1,+1 represents the modes and confidence level from A-C where A means higher confidence.")
    })
    
    
    output$markers_display<-renderTable({
      if(input$action1==0){
        return()
      }
      pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)
      write.table(pb.markers,file="Markers_info.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      head(pb.markers)
      
    })
    
    
     #coding for the tissue enrich 
    
    tissue<-reactive({
      clus_assay<-GetAssayData(pb)
      
      clus_assay<-as.data.frame(clus_assay)
      
      write.csv(clus_assay,file="clus_assay.csv",row.names = TRUE)
    })
    
    cluster_ident_ <-reactive({
      cluster_ident<-pb@active.ident
      
      cluster_ident<-as.data.frame(cluster_ident)
     # names(cluster_ident)[1] <- "Sample"
      write.csv(cluster_ident,file="cluster_ident.csv",row.names = TRUE)
      cluster_ident
    })
    
    output$cluster_ident<- renderPrint(
      {
        if(input$action1==0){
          return()
        }
        tissue()####### after added
        cluster_ident_()
        
        
      })
    output$downloadData_cluster_ident <- downloadHandler(
      filename = function() {
        paste("data-", Sys.Date(), ".csv", sep="")
      },
      content = function(file) {
        write.csv(cluster_ident_(), file) 
        
      }) 
    
    
    output$downloadData_meta <- downloadHandler(
      filename = function() {
        paste("data-", Sys.Date(), ".csv", sep="")
      },
      content = function(file) {
        write.csv(meta_file(), file) 
        
      }) 
    
    marker_table<- reactive({
      ################# adding for tissue enrich
      
      marker_dataframe <- subset(pb.markers,pb.markers$cluster == input$obs )
      
      gene_name_cluster<- marker_dataframe$gene
      write.table(gene_name_cluster,"gene_name_cluster.txt",row.names=F)
      
      marker_dataframe
      
    })
    marker_table_grn<- reactive({
      ################# adding for tissue enrich
      
      marker_dataframe <- subset(pb.markers,pb.markers$cluster == input$grn_obs )
      
      gene_name_cluster<- marker_dataframe$gene
      write.table(gene_name_cluster,"gene_name_cluster.txt",row.names=F)
      
      marker_dataframe
      
    })
 
    output$cluster_name<-renderPrint({
      if(input$action24==0){
        return() 
      }
      marker_table()
      
      # write.csv(marker_dataframe,"marker_dataframe.csv"
      
    })
    
    output$grn_clust<-renderPrint({
      if(input$action1==0){
        return() 
      }
      marker_table_grn() 
    })
    
    output$downloadData_cluster_name <- downloadHandler(
      filename = function() {
        paste("data-", Sys.Date(), ".csv", sep="")
      },
      content = function(file) {
        write.csv(marker_table(), file) 
        
      }) 
   # background_list <- reactive({
    
    # file <- input$backgorund
    #  if(is.null(file)){return()} 
    # read.table(file=file$datapath )
    
    #})
    tissue_enrich<-reactive({
      
      
      #gene_name_cluster<- marker_dataframe$gene
      
      
      
      inputGenes<-scan("gene_name_cluster.txt",character())
      inputGenes
      
    })
    
    output$tissue_detail<-renderPrint({
      if(input$action1==0){
        return()
      }
     #tissue_enrich()
      marker_table()
     inputGenes <-scan("gene_name_cluster.txt",character())
     inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
     
       write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
       seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
    
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
     # tissue_subset <-as.data.frame(tissue_subset)
      tab<- enrichmentOutput[,-4]
      tab
      log_df<-subset(tab, tab$Log10PValue!=0.00000000 & tab$fold.change!=0.00000000)
      log_df
      
    })
  
    
    output$downloadtissuedata <- downloadHandler(
      filename = function() {
        paste("data-", Sys.Date(), ".csv", sep="")
      },
      content = function(file) {
        write.csv(enrichmentOutput, file) 
        
      }) 
    
    plot_fold_t<-reactive({
      #tissue_enrich()
      #inputGenes<-scan("gene_name_cluster.txt",character())
      
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      tab<- enrichmentOutput[,-4]
      tab
      log_df<-subset(tab, tab$Log10PValue!=0.00000000 & tab$fold.change!=0.00000000)
      log_df
      
      ggplot(log_df,aes(x=reorder(Tissue,-fold.change),y=fold.change,label = Tissue.Specific.Genes,fill = Tissue))+
        geom_bar(stat = 'identity', width=0.5)+
        labs(x='', y = 'Fold change')+
        theme_bw()+
        theme(legend.position="none")+
        theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
   
       })
    
    output$plot_fold<-renderPlot({
      if(input$action1==0){
        return()
      }
      #for_plot()
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      tab<- enrichmentOutput[,-4]
      tab
      log_df<-subset(tab, tab$Log10PValue!=0.00000000 & tab$fold.change!=0.00000000)
      log_df
      
      ggplot(log_df,aes(x=reorder(Tissue,-fold.change),y=fold.change,label = Tissue.Specific.Genes,fill = Tissue))+
        geom_bar(stat = 'identity', width=0.5)+
        labs(x='', y = 'Fold change')+
        theme_bw()+
        theme(legend.position="none")+
        theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
   
      
    })
    
    output$tissue_fold1 <- downloadHandler(
      filename = function(){'fold_change.pdf'},
      content = function(file){
        pdf(file)
        print(  
          plot_fold_t()
         
        )
        dev.off()
      })
    
    
    output$tissue_log1 <- downloadHandler(
      filename = function(){'log_pvalue.pdf'},
      content = function(file){
        pdf(file)
        print(  
          
          logfold()
        
          
        )
        dev.off()
      })
    
    output$plot_log<-renderPlot({
      if(input$action1==0){
        return()
      }
     # tissue_enrich()
      #inputGenes<-scan("gene_name_cluster.txt",character())
      
      
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      tab<- enrichmentOutput[,-4]
      tab
      log_df<-subset(tab, tab$Log10PValue!=0.00000000 & tab$fold.change!=0.00000000)
      log_df
      ggplot(log_df,aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
        geom_bar(stat = 'identity', width=0.5)+
        labs(x='', y = '-LOG10(P-Adjusted)')+
        theme_bw()+
        theme(legend.position="none")+
        theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
        theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())
    })
    
    logfold<-reactive({ inputGenes <-scan("gene_name_cluster.txt",character())
    inputGenes
    
    gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
    output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
    output2
    enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
    
    write.table(enrichmentOutput,file="enrichment_output.csv",
                sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
    
    seEnrichmentOutput<-output2[[1]]
    
    enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
    
    
    enrichmentOutput$Tissue<-row.names(enrichmentOutput)
    
    
    #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
    
    
    # tissue_subset <-as.data.frame(tissue_subset)
    tab<- enrichmentOutput[,-4]
    tab
    log_df<-subset(tab, tab$Log10PValue!=0.00000000 & tab$fold.change!=0.00000000)
    log_df
    ggplot(log_df,aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,label = Tissue.Specific.Genes,fill = Tissue))+
      geom_bar(stat = 'identity', width=0.5)+
      labs(x='', y = '-LOG10(P-Adjusted)')+
      theme_bw()+
      theme(legend.position="none")+
      theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title = element_text(size=15))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),panel.grid.major= element_blank(),panel.grid.minor = element_blank())})
    
   output$kay_name<-renderPrint({
     inputGenes <-scan("gene_name_cluster.txt",character())
     inputGenes
     
     gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
     output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
     output2
     enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
     
     write.table(enrichmentOutput,file="enrichment_output.csv",
                 sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
     
     seEnrichmentOutput<-output2[[1]]
     
     enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
     
     
     enrichmentOutput$Tissue<-row.names(enrichmentOutput)
     
     
     #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
     
     
     # tissue_subset <-as.data.frame(tissue_subset)
     
     enrichmentOutput
     
     seEnrichmentOutput<-output2[[1]]
     
     enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
     
     enrichmentOutput$Tissue<-row.names(enrichmentOutput)
     
     seGroupInf<-output2[[3]][[input$kar1]]
    groupInf<-data.frame(assay(seGroupInf))
    #xy.list <- as.list(as.data.frame(groupInf))
 #  s1 <- groupInf$Gene
    
    groupInf$Gene
    
    
    })
   

    umap_plot<-reactive({
      DimPlot(pb, reduction = "umap",pt.size = 3)
      
    })
    
    output$umap<-renderPlot({
      if(input$action1==0){
        return()
      }
      pb <<- FindNeighbors(pb, dims = 1:10)
      pb <<- FindClusters(pb, resolution = 0.5)
      
      head(Idents(pb), 5)
      pb <<- RunUMAP(pb, dims = 1:10)
      DimPlot(pb, reduction = "umap",pt.size = 3)
      
    })
    output$umapPlot <- downloadHandler(
      filename = function(){'UMAP_plot.pdf'},
      content = function(file){
        pdf(file)
        print(      
          umap_plot()
          
                     
        )       
        dev.off()
      })
    
    list<-reactive({
      lst1=list()
      for (i in a){lst[[i]]=assay(seGroupInf[[i]])}
      lst1
    })
    pbmarker<-reactive({
      pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)
      
    })
    
    output$markers<-renderPrint({
      if(input$action1==0){
        return()
      }
      pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)
       write.table(pb.markers,file="Markers_info.csv",
                 sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
       print("Markers identified!!!!!")
    })
    
    output$downloadData <- downloadHandler(
      filename = function() {
        paste("data-", Sys.Date(), ".csv", sep="")
      },
      content = function(file) {
        write.csv(pb.markers, file) 
        
      }) 
     
    

    
   
    
   output$stouffer_sc_plot<-renderPlot({
      if(input$action1==0){
        return()
      }
     else if(input$st_sc =="Devloping heart atrial cardiomyocyte"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Devloping heart atrial cardiomyocyte.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     else if(input$st_sc =="Embryonic_stem_cell"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Embryonic_stem_cell.tsv.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     else if(input$st_sc =="Devloping heart ventricular cardiomyocyte"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Devloping heart ventricular cardiomyocyte.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     
     else if(input$st_sc =="Ectodermal cell differentiation"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Ectodermal cell differentiation.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     else if(input$st_sc =="Ectodermal development.tsv"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Ectodermal development.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     
     else if(input$st_sc =="Endodermal cell differentiation"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Endodermal cell differentiation.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     else if(input$st_sc =="Endodermal development"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Endodermal development.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     
     else if(input$st_sc =="Myofibroblast differentiation"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Myofibroblast differentiation.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     
     
     else if(input$st_sc =="Mesoderm development"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Mesoderm development.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     else if(input$st_sc =="Mesodermal cell differentiation"){
       
       matric<-pb@assays$RNA@counts
       
       cells_sum <- Matrix::colSums(matric>=3)
       
       matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
       
       matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
       
       cells_sum<-Matrix::rowSums(t(matric))
       
       matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
       r_name<-rownames(matric)
       media_genes <- read.table("Mesodermal cell differentiation.tsv")
       same_genes<-intersect(r_name,media_genes[,1])
       if(length(same_genes)==1)
         message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
       if(length(same_genes)==0)
         message="no common gene "
       if(length(same_genes)>1){
         matric= matric[same_genes,]####### didnot work why?
         
         #Further insuring have cells with atleast 10% expressed genes
         matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
         print(dim(matric))
         seurat_RNA_matric<-matric
         print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
         print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
         print(dim(seurat_RNA_matric))
         ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
         seurat_RNA_matric[seurat_RNA_matric==0]=1
         seurat_RNA_matric<-log2(seurat_RNA_matric)
         seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
         ###Stouffer score and then boxplot for each cell type
         stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
         print(sum(is.na(stouffer_score)))
         cell_types<-Idents(pb)[names(stouffer_score)]
         unique_cell_types=unique(cell_types)
         Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
         ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
           geom_boxplot(position=position_dodge(.2)) +
           geom_jitter(shape=16, position=position_jitter(.1)) +
           theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
     
     

else if(input$st_sc =="Adult chemosensory "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("final_list.csv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc=="Adult adrenal gland inflammatory cell")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult adrenal gland inflammatory cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal B cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal B cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$sc_st=="Fetal Basophil_Mast")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Basophil_Mast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CD1C+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CD1C+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CEPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CEPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CLEC9A+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CLEC9A+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Collecting duct lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Collecting duct lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Connecting tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Connecting tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Distal tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Distal tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal EBMPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal EBMPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal EEPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal EEPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Enteroendocrine cells_intestine_pancreas")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Enteroendocrine cells_intestine_pancreas.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Enteroendocrine cells_stomach")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Enteroendocrine cells_stomach.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Erythroblast")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Erythroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal ETDs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal ETDs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal HSCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal HSCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal HSPCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal HSPCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal IL1B+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal IL1B+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal ILC 3")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal ILC 3.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Islet beta cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Islet beta cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Islet delta cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Islet delta cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal LEC")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal LEC.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Loop of Henle lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Loop of Henle lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Meg progenitors")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Meg progenitors.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Meg")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Meg.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Megakaryoblasts")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Megakaryoblasts.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Nephron progenitor cell")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Nephron progenitor cell")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal NK cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal NK cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal pDCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal pDCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Perivascular macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Perivascular macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Phagocytic macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Phagocytic macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Plasma cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Plasma cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Podocyte lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Podocyte lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Proximal tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Proximal tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal PTPRC+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal PTPRC+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Pulmonary neuroendocrine cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Pulmonary neuroendocrine cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Renal vesicle")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Renal vesicle.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$sc_st=="Fetal S100A9+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal S100A9+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal T cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal T cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal TMEM119+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal TMEM119+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal TRAF1+ APCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal TRAF1+ APCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Ureteric tip")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Ureteric tip.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal adrenal")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal adrenal.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal brain")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal brain.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal kidney")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal kidney.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal liver")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal liver.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal lung")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal lung.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal placenta")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal placenta.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal spleen")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal spleen.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult antigen presenting cell (RPS high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult antigen presenting cell (RPS high).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult astrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult astrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult AT2_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult AT2 cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult b_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult b_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult b_cell_plasmocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult b_cell_plasmocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult basal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult basal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult CB_CD34+"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult CB CD34+23.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult chondrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult chondrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult dendritic_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult dendritic_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial cell (APC)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial cell (APC).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial cell (endothelial to mesenchymal transition)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial cell (endothelial to mesenchymal transition).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult enterocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult enterocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult enterocyte_progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult enterocyte_progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult epithelial cell (intermediated)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult epithelial cell (intermediated).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult epithelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult epithelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult erythroid cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult erythroid cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult erythroid progenitor cell (RP high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult erythroid progenitor cell (RP high).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult fasciculata_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult fasciculata_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal acinar cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal acinar cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal chondrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal chondrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Fetal endocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal endocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal enterocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal enterocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Fetal epithelial progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal epithelial progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal fibroblast"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal fibroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal skeletal muscle cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal skeletal muscle cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_mesenchymal_progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_mesenchymal_progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_neuron "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_neuron .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_stromal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_stromal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult Stromal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult stromal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult fibroblast"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult fibroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult gastric chief cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult gastric chief cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult gastric endocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult gastric endocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult goblet_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult goblet_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult hepatocyte_Endodermal cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult hepatocyte_Endodermal cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult hESC87"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult hESC87.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult immature sertoli cell (Pre-Sertoli cell)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult immature sertoli cell (Pre-Sertoli cell).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult intercalated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult intercalated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult intermediated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult intermediated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult kidney intercalated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult kidney intercalated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult loop of Henle"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult loop of Henle.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult M2 Macrophage"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult M2 Macrophage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult macrophage"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult macrophage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult mast cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult mast cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult mesothelial cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult mesothelial cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult monocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Monocyte.csv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult myeloid cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult myeloid cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Neutrophil (RPS high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Neutrophil (RPS high)40.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult neutrophil"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult neutrophil.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult oligodendrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult oligodendrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult pancreas exocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult pancreas exocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult primordial germ cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult primordial germ cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult proliferating T cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult proliferating T cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult proximal tubule progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult proximal tubule progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult sinusoidal endothelial cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult sinusoidal endothelial cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult smooth_muscle_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult smooth_muscle_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult stratified_epithelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult stratified_epithelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult T_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult T_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult thyroid follicular cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult thyroid follicular cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult ureteric bud cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("dult ureteric bud cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult ventricle cardiomyocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult ventricle cardiomyocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


     
     
     
    })
   
    
    
    
    
    
    
    
    
    
    stuouffer<-reactive({
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- Read_csv()
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) }
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))
    })
    
    
    
    
    
    
    
    
    
    
    
    
    output$stouffer_table<- renderPrint({
      if(input$action230==0){
        return()
      }
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- Read_csv()
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      
      print(sum(is.na(stouffer_score)))
      
      cell_types<-Idents(pb)[names(stouffer_score)]
      
      unique_cell_types=unique(cell_types)
      
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types)}
      
      Stouffer_score_df
      
    })
      
    output$de_stouffer<-renderPlot({
      if(input$action230==0){
        return()
      }
      stuouffer()
    })
    
    output$stoufferPlot <- downloadHandler(
      filename = function(){'Stouffer_plot.pdf'},
      content = function(file){
        pdf(file)
        print(       
          ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
            geom_boxplot(position=position_dodge(.2)) +
            geom_jitter(shape=16, position=position_jitter(.1)) +
            theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))+
            scale_fill_brewer(palette="Dark2")     
                     
        )
        dev.off()
      })
    
    output$stouffer_pval<-renderPrint({
      if(input$action2300==0){
        return()
      }
      matric<-pb@assays$RNA@counts
      cells_sum <- Matrix::colSums(matric>=3)
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      cells_sum<-Matrix::rowSums(t(matric))
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      
      media_genes <- Read_csv()
      common_genes=intersect(rownames(matric),media_genes[,1])
      matric= matric[common_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      
      print(sum(is.na(stouffer_score)))
      
      cell_types<-Idents(pb)[names(stouffer_score)]
      
      unique_cell_types=unique(cell_types)
      
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types)
      p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
      colnames(p_val_stouffer_score)=unique_cell_types
      rownames(p_val_stouffer_score)=unique_cell_types
      for (i in 1:dim(p_val_stouffer_score)[1])
      {
        for (j in 1:dim(p_val_stouffer_score)[2])
        {
          print(paste(i,"_",j,sep=""))
          a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
          b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
          p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
        }
      }
      format(round(p_val_stouffer_score, 4))
    })
    
    output$de_karyotype<-renderPlot({
      if(input$action1==0){
        return()
      }
      
      library(TxDb.Hsapiens.UCSC.hg19.knownGene)
      txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
      dm.genes <- genes(txdb)
      dm.genes
      
      library("org.Hs.eg.db")
      kar<-pb.markers
      gene<-(input$detected_tissue)
      gene
      gene_id<-unlist(mget(x=gene,envir=org.Hs.egALIAS2EG))
      gene_id
      
      #Merging the data frames
      gened<-as.data.frame(gene_id)
      #gened$name<-row.names(gened)
      askk<-data.frame("gene"=rownames(gened))
      askk$p_val<-kar[gene,]$p_val
      askk$avg_log2FC<-kar[gene,]$avg_log2FC
      askk$pct.1<-kar[gene,]$pct.1
      askk$pct.2<-kar[gene,]$pct.2
      askk$p_val_adj<-kar[gene,]$p_val_adj
      askk$cluster<-kar[gene,]$cluster
      askk$gene_id<-gened$gene_id
      rownames(askk)<-askk$gene_id
      
      
      
      #ask<-merge(kar,gened,by.x="gene",by.y="name")
      
      ##Final data frame with start and end
      mcols(dm.genes) <- askk[names(dm.genes), c("avg_log2FC","p_val","p_val_adj","cluster","gene")]
      head(dm.genes, n=4)
      
      ##
      library(karyoploteR)
      filtered.dm.genes_mat <- dm.genes[!is.na(dm.genes$p_val_adj)]
      filtered.dm.genes<-filtered.dm.genes_mat[gened$gene_id,]
      ord<-as.data.frame(filtered.dm.genes)
      rownames(ord)<-ord$gene
      log.pval <- -log10(filtered.dm.genes$p_val_adj)
      mcols(filtered.dm.genes)$log.pval <- log.pval
      filtered.dm.genes
      
      range(filtered.dm.genes$avg_log2FC)
      fc.ymax <- ceiling(max(abs(range(filtered.dm.genes$avg_logFC))))
      fc.ymin <- -fc.ymax
      
      cex.val <- sqrt(filtered.dm.genes$log.pval)/2
      
      col.over <- "#00A6EDAA"
      sign.col <- rep(col.over, length(filtered.dm.genes))
      chrs<-unique(ord$seqnames)
      
      ##Final Ploting
      kp <- plotKaryotype(genome="hg19")
      #kpPoints(kp, data=filtered.dm.genes, y=filtered.dm.genes$avg_log2FC,  cex=cex.val, ymax=fc.ymax, ymin=fc.ymin,r1=0.1,col=sign.col)
      #kpAxis(kp, ymax=fc.ymax, ymin=fc.ymin,r1=0.8)
      #kpPlotMarkers(kp, filtered.dm.genes, labels = rownames(ord), text.orientation = "horizontal", ignore.chromosome.ends=TRUE, r0=0.01,marker.parts=c(0.1,0.1,0.1),label.dist=0.03)
      #kpAddLabels(kp, labels = "log2FC", srt=90, pos=1, label.margin = 0.04, ymax=fc.ymax, ymin=fc.ymin,r1=0.8)
      #gene.mean <- start(filtered.dm.genes) + (end(filtered.dm.genes) - start(filtered.dm.genes))/2
      #kpSegments(kp, chr=as.character(seqnames(filtered.dm.genes)), x0=gene.mean, x1=gene.mean, y0=filtered.dm.genes$avg_log2FC, y1=fc.ymax, ymax=fc.ymax, ymin=fc.ymin, r1=0.8)
      
      ##Selected Genes
      gen<-filtered.dm.genes$gene
      
      # gene_list<- c(input$detected_tissue)
      col.under <- "#0000FF"
      sign.col[filtered.dm.genes$gene==gen] <- col.under
      kpPoints(kp, data=filtered.dm.genes, y=filtered.dm.genes$avg_log2FC, ymax=fc.ymax, ymin=fc.ymin,r1=0.1,col=sign.col)
      #selected.dm.genes <- filtered.dm.genes[filtered.dm.genes$gene]
      kpPlotMarkers(kp, filtered.dm.genes, labels = gen, text.orientation = "horizontal", ignore.chromosome.ends=TRUE, r0=0.01,marker.parts=c(0.1,0.1,0.1),label.dist=0.005,label.color =  "#0000FF")
      
      
      
      
      
      
      
    })
    output$karyoplott <- downloadHandler(
      filename = function(){'karyoplot_plot.pdf'},
      content = function(file){
        pdf(file)
        print(    kayro())
        
        dev.off()
      })
    
  kayro<-reactive({
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
    dm.genes <- genes(txdb)
    dm.genes
    
    library("org.Hs.eg.db")
    kar<-pb.markers
    gene<-(input$detected_tissue)
    gene
    gene_id<-unlist(mget(x=gene,envir=org.Hs.egALIAS2EG))
    gene_id
    
    #Merging the data frames
    gened<-as.data.frame(gene_id)
    #gened$name<-row.names(gened)
    askk<-data.frame("gene"=rownames(gened))
    askk$p_val<-kar[gene,]$p_val
    askk$avg_log2FC<-kar[gene,]$avg_log2FC
    askk$pct.1<-kar[gene,]$pct.1
    askk$pct.2<-kar[gene,]$pct.2
    askk$p_val_adj<-kar[gene,]$p_val_adj
    askk$cluster<-kar[gene,]$cluster
    askk$gene_id<-gened$gene_id
    rownames(askk)<-askk$gene_id
    
    
    
    #ask<-merge(kar,gened,by.x="gene",by.y="name")
    
    ##Final data frame with start and end
    mcols(dm.genes) <- askk[names(dm.genes), c("avg_log2FC","p_val","p_val_adj","cluster","gene")]
    head(dm.genes, n=4)
    
    ##
    library(karyoploteR)
    filtered.dm.genes_mat <- dm.genes[!is.na(dm.genes$p_val_adj)]
    filtered.dm.genes<-filtered.dm.genes_mat[gened$gene_id,]
    ord<-as.data.frame(filtered.dm.genes)
    rownames(ord)<-ord$gene
    log.pval <- -log10(filtered.dm.genes$p_val_adj)
    mcols(filtered.dm.genes)$log.pval <- log.pval
    filtered.dm.genes
    
    range(filtered.dm.genes$avg_log2FC)
    fc.ymax <- ceiling(max(abs(range(filtered.dm.genes$avg_logFC))))
    fc.ymin <- -fc.ymax
    
    cex.val <- sqrt(filtered.dm.genes$log.pval)/2
    
    col.over <- "#00A6EDAA"
    sign.col <- rep(col.over, length(filtered.dm.genes))
    chrs<-unique(ord$seqnames)
    
    ##Final Ploting
    kp <- plotKaryotype(genome="hg19")
    #kpPoints(kp, data=filtered.dm.genes, y=filtered.dm.genes$avg_log2FC,  cex=cex.val, ymax=fc.ymax, ymin=fc.ymin,r1=0.1,col=sign.col)
    #kpAxis(kp, ymax=fc.ymax, ymin=fc.ymin,r1=0.8)
    #kpPlotMarkers(kp, filtered.dm.genes, labels = rownames(ord), text.orientation = "horizontal", ignore.chromosome.ends=TRUE, r0=0.01,marker.parts=c(0.1,0.1,0.1),label.dist=0.03)
    #kpAddLabels(kp, labels = "log2FC", srt=90, pos=1, label.margin = 0.04, ymax=fc.ymax, ymin=fc.ymin,r1=0.8)
    #gene.mean <- start(filtered.dm.genes) + (end(filtered.dm.genes) - start(filtered.dm.genes))/2
    #kpSegments(kp, chr=as.character(seqnames(filtered.dm.genes)), x0=gene.mean, x1=gene.mean, y0=filtered.dm.genes$avg_log2FC, y1=fc.ymax, ymax=fc.ymax, ymin=fc.ymin, r1=0.8)
    
    ##Selected Genes
    gen<-filtered.dm.genes$gene
    
    # gene_list<- c(input$detected_tissue)
    col.under <- "#0000FF"
    sign.col[filtered.dm.genes$gene==gen] <- col.under
    kpPoints(kp, data=filtered.dm.genes, y=filtered.dm.genes$avg_log2FC, ymax=fc.ymax, ymin=fc.ymin,r1=0.1,col=sign.col)
    #selected.dm.genes <- filtered.dm.genes[filtered.dm.genes$gene]
    kpPlotMarkers(kp, filtered.dm.genes, labels = gen, text.orientation = "horizontal", ignore.chromosome.ends=TRUE, r0=0.01,marker.parts=c(0.1,0.1,0.1),label.dist=0.005,label.color =  "#0000FF")
    
    
    
    
  })
    
    
    
   
    
    
    output$f_or<-renderUI({
      or<-read.csv("final_list.csv")
      row_names<-row.names(pb)
      row_names<-as.data.frame(row_names)
      colnames(row_names)<-"Receptor_names"
      factor0<<-merge(row_names,or,by.x="Receptor_names",by.y="Symbol")
      factor0
      selectInput("detected_or","Select Receptor",choices = factor0)
      #paste("Number of ORs detected",nrow(factor0))
    })
    
    output$or_umap<-renderPlot({
      FeaturePlot(pb, features = c(input$detected_or))
    })
    
    
    output$orPlot <- downloadHandler(
      filename = function(){'OR_plot.pdf'},
      content = function(file){
        pdf(file)
        print(      
          FeaturePlot(pb, features = c(input$detected_or))
          
          
        )       
        dev.off()
      })
  
    
    output$t_or<-renderUI({
      if(input$action1==0){
        return()
      }
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      
      enrichmentOutput
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      zzz<-  enrichmentOutput$Tissue
      zzz
     
      
      selectInput("name_tissue","Select Tissue",choices = c(zzz))
      
     
    })
    
    gene_tissue_detail<-reactive({
      #Retrieval of input tissue-specific genes
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      
      enrichmentOutput
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
     
      seGroupInf<-output2[[3]][[input$name_tissue]]
      groupInf<-data.frame(assay(seGroupInf))
      write.csv(groupInf,"tissue_sp_gene.csv")
      
    })
   output$tt_or<-renderUI({
     if(input$action1==0){
       return()
     }
    # gene_tissue_detail()

    inputGenes <-scan("gene_name_cluster.txt",character())
     inputGenes
     
     gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
     output2<-teEnrichment(gs, rnaSeqDataset = input$dataset)#here
     output2
     enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
     
     write.table(enrichmentOutput,file="enrichment_output.csv",
                 sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
     
     seEnrichmentOutput<-output2[[1]]
     
     enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
     
     
     enrichmentOutput$Tissue<-row.names(enrichmentOutput)
     
     
     #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
     
     
     # tissue_subset <-as.data.frame(tissue_subset)
     
     enrichmentOutput
     
     seEnrichmentOutput<-output2[[1]]
     
     enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
     
     enrichmentOutput$Tissue<-row.names(enrichmentOutput)
     seGroupInf<-output2[[3]][[input$name_tissue]]
     groupInf<-data.frame(assay(seGroupInf))
     write.csv(groupInf,"tissue_sp_gene.csv")
     ttor<- read.csv("tissue_sp_gene.csv")
     # ttor<-as.data.frame(ttor)
     #ttor<-ttor$Gene
     row_names1<-row.names(pb)
     row_names1<-as.data.frame(row_names1)
    colnames(row_names1)<-"Receptor_names"
     factor100<<-merge(row_names1,ttor,by.x="Receptor_names",by.y="Gene")
     factor1000<-factor100$Receptor_names
     factor1000
     
     selectInput("detected_tissue","Select Transcript",choices = factor1000)
     #paste("Number of ORs detected",nrow(factor0)) 
   }) 
    
    output$tor_umap<-renderPlot({
      FeaturePlot(pb, features = c(input$detected_tissue))
    })
    
    
    output$torPlot <- downloadHandler(
      filename = function(){'umap_plot.pdf'},
      content = function(file){
        pdf(file)
        print(      
          FeaturePlot(pb, features = c(input$detected_tissue))
          
          
        )       
        dev.off()
      })
    
    
    
  #  output$other_or<-renderUI({
   #   other_or<-read.csv("other_receptors.csv")
    #  other_receptors<-row.names(pb)
     # other_receptors<-as.data.frame(other_receptors)
      #colnames(other_receptors)<-"Receptor_names"
      #factor_other<<-merge(other_receptors,other_or,by.x="Receptor_names",by.y="Symbol")
      #selectInput("detected_other_or","Select Receptor",choices = factor_other)
      #paste("Number of ORs detected",nrow(factor0))
    #})
   # 
    #output$other_umap<-renderPlot({
     # FeaturePlot(pb, features = c(input$detected_other_or))
    #})
    
    output$user_umap<-renderPlot({
      FeaturePlot(pb, features = c(input$usergene))
    })
    output$uorPlot <- downloadHandler(
      filename = function(){'UOR_plot.pdf'},
      content = function(file){
        pdf(file)
        print(      
          FeaturePlot(pb, features = c(input$usergene))
          
          
        )       
        dev.off()
      })  
    
    
    
    
    output$grn<- renderTable({
      if(input$action6000==0){
        return()
      }
   
      
      pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc )### have to make chnage in threshold
      
      ##########************************* read dorothea_regulon file
      dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
      
      ## We obtain the regulons based on interactions with confidence level A, B and C
      regulon <- dorothea_regulon_human %>%
        dplyr::filter(confidence %in% c("A","B","C")) 
      
      ## We compute Viper Scores 
      pb <- run_viper(pb, regulon,
      )
      
      ## We compute the Nearest Neighbours to perform cluster
      DefaultAssay(object = pb) <- "dorothea"
      pb <- ScaleData(pb)
      pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
      pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
      pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
      
      pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
      
      pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                                   logfc.threshold = input$log_fc, verbose = FALSE)
      
      DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
      
      ## We transform Viper scores, scaled by seurat, into a data frame to better 
      
      ## handling the results
      viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                      assay = "dorothea") %>%
        data.frame() %>%
        t()
      
      ## We create a data frame containing the cells and their clusters
      CellsClusters <- data.frame(cell = names(Idents(pb)), 
                                  cell_type = as.character(Idents(pb)),
                                  stringsAsFactors = FALSE)
      
      
      ################## my addition ###################################
      viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
      viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
      viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
      
      summarized_viper_scores <- viper_scores_clusters %>% 
        group_by(variable, cell_type) %>%
        summarise(avg = mean(value),
                  std = sd(value))
      ## We select the 20 most variable TFs. (20*9 populations = 180)
      
      highly_variable_tfs <- summarized_viper_scores %>%
        group_by(variable) %>%
        mutate(var = var(avg))  %>%
        ungroup() %>%
        top_n(180, var) %>%
        distinct(variable)
      
      ## We prepare the data for the plot
      summarized_viper_scores_df <- summarized_viper_scores %>%
        semi_join(highly_variable_tfs, by = "variable") %>%
        dplyr::select(-std) %>%   
        spread(variable, avg) %>%
        data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
      
      transpose_summarise_df<-t(summarized_viper_scores_df)
      
      merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
      final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
      
      inputGenes <-scan("gene_name_cluster.txt",character())
      inputGenes
      
      gs<-GeneSet(geneIds=inputGenes,organism="Homo Sapiens",geneIdType=SymbolIdentifier())# changes TE custom to te
      output2<-teEnrichment(gs)#here
      output2
      enrichmentOutput<<-setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
      
      write.table(enrichmentOutput,file="enrichment_output.csv",
                  sep=",",row.names=FALSE,col.names = TRUE,quote = FALSE)
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      
      #tissue_subset <-subset(enrichmentOutput$Tissue, enrichmentOutput$fold.change!= 0)
      
      
      # tissue_subset <-as.data.frame(tissue_subset)
      
      enrichmentOutput
      
      seEnrichmentOutput<-output2[[1]]
      
      enrichmentOutput<-setNames(data.frame(assay(seEnrichmentOutput), row.names = rowData(seEnrichmentOutput)[,1]), colData(seEnrichmentOutput)[,1])
      
      enrichmentOutput$Tissue<-row.names(enrichmentOutput)
      
      seGroupInf<-output2[[3]]
      m=matrix(0,nrow=35,ncol= 20)
      rownames(m)=names(seGroupInf)
      a=names(seGroupInf)
      
      for (i in a){if(length(assay(seGroupInf[[i]])[,"Gene"])!=0)m[i,1:length(assay(seGroupInf[[i]])[,"Gene"])]=assay(seGroupInf[[i]])[,"Gene"]}
      m<-as.data.frame(m)
      names(m)[1:10]<- "target"
      tm<-t(m)
      melted_m<- melt(tm,id="tissuenames",measure.vars = c("target","target.1","target.2","target.3","target.4","target.5","target.6","target.7","target.8","target.9","target.10"))
      melted_m<-unique(melted_m)
      updated_myData <- subset(melted_m, value!= 0)
      finaldata_tissuenerich<-merge(final_data_plot,updated_myData,by.x="target", by.y="value")
      colnames(finaldata_tissuenerich)[which(names(finaldata_tissuenerich) == "Row.names")] <- "TF"
      
      final_data_toshow<- finaldata_tissuenerich  %>% dplyr::select(target,Var2,TF,mor)
      colnames(final_data_toshow)[which(names(final_data_toshow) == "Var2")] <- "Tissue"
      colnames(final_data_toshow)[which(names(final_data_toshow) == "target")] <- "Target"
      
      final_data_toshow
      #chordDiagram(finaldata_tissuenerich)
      #final_data_toshow$mor <- as.factor(final_data_toshow$mor)
      
     # ggplot(as.data.frame(final_data_toshow),
     #        aes(axis1 = Tissue, axis2 = TF, axis3 = Target)) +
     ###   geom_alluvium(aes(fill = mor), width = 1/12) +
       ## geom_stratum(width = 1/12, fill = "black", color = "grey") +
       # scale_x_continuous(breaks = 1:3, labels = c("Tissue", "TF", "Target")) +
        #scale_fill_brewer(type = "qual", palette = "Set1") +
        #ggtitle("Gene regulatory network") + 
        #coord_flip() + 
        #ggfittext::geom_fit_text(aes(label = after_stat(stratum)), reflow = TRUE, stat = "stratum", width = 1/4, min.size = 1, colour = "white", size = 10) +
        #theme_minimal()
     
      
     
    })
      # alluvial::alluvial(finaldata_tissuenerich,freq = dim(finaldata_tissuenerich),col = "blue")
      
      #chordDiagram(finaldata_tissuenerich)
  
      
      
      
    
    output$grnPlot <- downloadHandler(
      filename = function(){'GRN.pdf'},
      content = function(file){
        pdf(file)
        print(   plot(plotg, layout=-l$layout[,2:1],vertex.size = 25, vertex.color = "PINK",
                      vertex.frame.color = "RED", vertex.label.color = "black",
                      vertex.label.cex = 1)    
                 
                 
        )       
        dev.off()
      })
    
    
    AUCell_table<-reactive({
      au_mat<-pb@assays$RNA@counts
      au_mat<-as.matrix(au_mat)
      
      cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
      cells_rankings
      save(cells_rankings, file="cells_rankings.RData")
      cells_AUC <- AUCell_calcAUC(abc[[input$aucell]], cells_rankings)
      save(cells_AUC, file="cells_AUC.RData")    
      set.seed(123)
      par(mfrow=c(1,1)) 
      cells_assignment <- AUCell_exploreThresholds(cells_AUC, assign=TRUE) 
      score_cell<-as.matrix(cells_assignment)
   score_cell
     
    })
    AUCell<-reactive({
      au_mat<-pb@assays$RNA@counts
      au_mat<-as.matrix(au_mat)
      
      cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
      cells_rankings
      save(cells_rankings, file="cells_rankings.RData")
      cells_AUC <- AUCell_calcAUC(abc[[input$aucell]], cells_rankings)
      save(cells_AUC, file="cells_AUC.RData")    
      set.seed(123)
      par(mfrow=c(1,1)) 
      cells_assignment <- AUCell_exploreThresholds(cells_AUC, assign=TRUE) 
      df2 <- as.data.frame(lapply(cells_assignment, unlist))
      sumByGene <- apply(au_mat, 1, sum)
      exprMatSubset <- au_mat[which(sumByGene>0),]
      logMatrix <- log2(exprMatSubset+1)
      
      umap_mat<-pb[["umap"]]@cell.embeddings
      umap_mat<-as.data.frame(umap_mat)
      selectedThresholds <- getThresholdSelected(cells_assignment)
      
      for(geneSetName in names(selectedThresholds))
      {
        nBreaks <- 5 # Number of levels in the color palettes
        # Color palette for the cells that do not pass the threshold
        colorPal_Neg <- grDevices::colorRampPalette(c("black","blue", "skyblue"))(nBreaks)
        # Color palette for the cells that pass the threshold
        colorPal_Pos <- grDevices::colorRampPalette(c("pink", "magenta", "red"))(nBreaks)
        
        # Split cells according to their AUC value for the gene set
        passThreshold <- getAUC(cells_AUC)[geneSetName,] >  selectedThresholds[geneSetName]
        if(sum(passThreshold) >0 )
        {
          aucSplit <- split(getAUC(cells_AUC)[geneSetName,], passThreshold)
          
          # Assign cell color
          cellColor <- c(setNames(colorPal_Neg[cut(aucSplit[[1]], breaks=nBreaks)], names(aucSplit[[1]])), 
                         setNames(colorPal_Pos[cut(aucSplit[[2]], breaks=nBreaks)], names(aucSplit[[2]])))
          
          # Plot
          plot(umap_mat, main=geneSetName,
               sub="Pink/red cells pass the threshold",
               col=cellColor[rownames(umap_mat)], pch=16) 
        }
      }
      
      
      
    })

      output$au_cell<-renderPlot({
    if(input$action1==0){
      return()
      }
        au_mat<-pb@assays$RNA@counts
        au_mat<-as.matrix(au_mat)
   
        cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
        cells_rankings
        save(cells_rankings, file="cells_rankings.RData")
        cells_AUC <- AUCell_calcAUC(abc[[input$aucell]], cells_rankings)
        save(cells_AUC, file="cells_AUC.RData")    
        set.seed(123)
        par(mfrow=c(1,1)) 
        cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=TRUE, assign=TRUE) 
        
         })  
  
      output$cell_assign_score <- downloadHandler(
        filename = function()
          { paste("data-", Sys.Date(), ".csv", sep="")
          },
        content = function(file){
           
          write.csv(AUCell_table(),file) 
          
          
        })
     
    
  
  
    output$cell_assign <- downloadHandler(
    filename = function(){'AUC.pdf'},
    content = function(file){
      pdf(file)
      print(      
       
        AUCell()        
      )       
      dev.off()
    })
  
  
  
  output$Stouffer_sc<-renderTable({
    if(input$action1==0){
      return()
    }
    else if(input$st_sc =="Devloping heart atrial cardiomyocyte"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Devloping heart atrial cardiomyocyte.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if(input$st_sc =="Embryonic_stem_cell"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Embryonic_stem_cell.tsv.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if(input$st_sc =="Devloping heart ventricular cardiomyocyte"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Devloping heart ventricular cardiomyocyte.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if(input$st_sc =="Ectodermal cell differentiation"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Ectodermal cell differentiation.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if(input$st_sc =="Ectodermal development.tsv"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Ectodermal development.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if(input$st_sc =="Endodermal cell differentiation"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Endodermal cell differentiation.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if(input$st_sc =="Endodermal development"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Endodermal development.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if(input$st_sc =="Myofibroblast differentiation"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Myofibroblast differentiation.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    
    else if(input$st_sc =="Mesoderm development"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Mesoderm development.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if(input$st_sc =="Mesodermal cell differentiation"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.table("Mesodermal cell differentiation.tsv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    

else if(input$st_sc =="Adult chemosensory "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("final_list.csv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc=="Adult adrenal gland inflammatory cell")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult adrenal gland inflammatory cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal B cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal B cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$sc_st=="Fetal Basophil_Mast")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Basophil_Mast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CD1C+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CD1C+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CEPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CEPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal CLEC9A+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal CLEC9A+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Collecting duct lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Collecting duct lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Connecting tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Connecting tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Distal tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Distal tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal EBMPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal EBMPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal EEPs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal EEPs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Enteroendocrine cells_intestine_pancreas")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Enteroendocrine cells_intestine_pancreas.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Enteroendocrine cells_stomach")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Enteroendocrine cells_stomach.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Erythroblast")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Erythroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal ETDs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal ETDs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal HSCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal HSCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal HSPCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal HSPCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal IL1B+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal IL1B+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal ILC 3")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal ILC 3.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Islet beta cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Islet beta cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Islet delta cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Islet delta cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal LEC")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal LEC.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Loop of Henle lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Loop of Henle lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Meg progenitors")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Meg progenitors.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Meg")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Meg.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Megakaryoblasts")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Megakaryoblasts.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Nephron progenitor cell")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Nephron progenitor cell")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal NK cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal NK cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal pDCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal pDCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Perivascular macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Perivascular macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Phagocytic macrophages")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Phagocytic macrophages.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Plasma cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Plasma cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Podocyte lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Podocyte lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Proximal tubule lineage")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Proximal tubule lineage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal PTPRC+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal PTPRC+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Pulmonary neuroendocrine cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Pulmonary neuroendocrine cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Renal vesicle")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Renal vesicle.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$sc_st=="Fetal S100A9+ DCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal S100A9+ DCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal T cells")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal T cells.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal TMEM119+ Microglia")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal TMEM119+ Microglia.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal TRAF1+ APCs")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal TRAF1+ APCs.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal Ureteric tip")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal Ureteric tip.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal adrenal")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal adrenal.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal brain")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal brain.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal kidney")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal kidney.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal liver")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal liver.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal lung")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal lung.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal placenta")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal placenta.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$sc_st=="Fetal spleen")
  
{
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Fetal spleen.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult antigen presenting cell (RPS high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult antigen presenting cell (RPS high).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult astrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult astrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult AT2_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult AT2 cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult b_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult b_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult b_cell_plasmocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult b_cell_plasmocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult basal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult basal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult CB_CD34+"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult CB CD34+23.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult chondrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult chondrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult dendritic_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult dendritic_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial cell (APC)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial cell (APC).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial cell (endothelial to mesenchymal transition)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial cell (endothelial to mesenchymal transition).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult endothelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult endothelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult enterocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult enterocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult enterocyte_progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult enterocyte_progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult epithelial cell (intermediated)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult epithelial cell (intermediated).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult epithelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult epithelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult erythroid cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult erythroid cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult erythroid progenitor cell (RP high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult erythroid progenitor cell (RP high).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult fasciculata_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult fasciculata_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal acinar cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal acinar cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal chondrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal chondrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Fetal endocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal endocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal enterocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal enterocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Fetal epithelial progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal epithelial progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal fibroblast"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal fibroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal skeletal muscle cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal skeletal muscle cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_mesenchymal_progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_mesenchymal_progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_neuron "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_neuron .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Fetal_stromal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Fetal_stromal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult Stromal_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult stromal_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult fibroblast"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult fibroblast.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult gastric chief cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult gastric chief cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult gastric endocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult gastric endocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult goblet_cell "){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult goblet_cell .tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult hepatocyte_Endodermal cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult hepatocyte_Endodermal cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult hESC87"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult hESC87.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult immature sertoli cell (Pre-Sertoli cell)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult immature sertoli cell (Pre-Sertoli cell).tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult intercalated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult intercalated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult intermediated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult intermediated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult kidney intercalated cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult kidney intercalated cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult loop of Henle"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult loop of Henle.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult M2 Macrophage"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult M2 Macrophage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult macrophage"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult macrophage.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult mast cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult mast cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult mesothelial cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult mesothelial cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult monocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Monocyte.csv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult myeloid cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult myeloid cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Neutrophil (RPS high)"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Neutrophil (RPS high)40.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult neutrophil"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.csv("Adult neutrophil.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult oligodendrocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult oligodendrocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult pancreas exocrine cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult pancreas exocrine cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult primordial germ cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult primordial germ cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult proliferating T cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult proliferating T cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult proximal tubule progenitor"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult proximal tubule progenitor.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult sinusoidal endothelial cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult sinusoidal endothelial cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult smooth_muscle_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult smooth_muscle_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult stratified_epithelial_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult stratified_epithelial_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

else if(input$st_sc =="Adult T_cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult T_cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult thyroid follicular cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult thyroid follicular cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}


else if(input$st_sc =="Adult ureteric bud cell"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("dult ureteric bud cell.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
else if(input$st_sc =="Adult ventricle cardiomyocyte"){
  
  matric<-pb@assays$RNA@counts
  
  cells_sum <- Matrix::colSums(matric>=3)
  
  matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
  
  matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
  
  cells_sum<-Matrix::rowSums(t(matric))
  
  matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
  r_name<-rownames(matric)
  media_genes <- read.table("Adult ventricle cardiomyocyte.tsv")
  same_genes<-intersect(r_name,media_genes[,1])
  if(length(same_genes)==1)
    message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
  if(length(same_genes)==0)
    message="no common gene "
  if(length(same_genes)>1){
    matric= matric[same_genes,]####### didnot work why?
    
    #Further insuring have cells with atleast 10% expressed genes
    matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
    print(dim(matric))
    seurat_RNA_matric<-matric
    print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
    print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
    print(dim(seurat_RNA_matric))
    ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
    seurat_RNA_matric[seurat_RNA_matric==0]=1
    seurat_RNA_matric<-log2(seurat_RNA_matric)
    seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
    ###Stouffer score and then boxplot for each cell type
    stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
    print(sum(is.na(stouffer_score)))
    cell_types<-Idents(pb)[names(stouffer_score)]
    unique_cell_types=unique(cell_types)
    Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
    ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
      geom_boxplot(position=position_dodge(.2)) +
      geom_jitter(shape=16, position=position_jitter(.1)) +
      theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

    p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
    colnames(p_val_stouffer_score)=unique_cell_types
    rownames(p_val_stouffer_score)=unique_cell_types
    for (i in 1:dim(p_val_stouffer_score)[1])
    {
      for (j in 1:dim(p_val_stouffer_score)[2])
      {
        print(paste(i,"_",j,sep=""))
        a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
        b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
        p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
      }
    }
    
    format(round(p_val_stouffer_score, 4))
    
  })
  
  
  
  
  output$au_score<-renderPrint({
    if(input$au1==0){
      return()
    }
    au_mat<-pb@assays$RNA@counts
    au_mat<-as.matrix(au_mat)
    par("mar")

    cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
    cells_rankings
    save(cells_rankings, file="cells_rankings.RData")
    cells_AUC <- AUCell_calcAUC(abc[[input$aucell]], cells_rankings)
    save(cells_AUC, file="cells_AUC.RData")    
    set.seed(123)
    par(mfrow=c(1,1)) 
    cells_assignment <- AUCell_exploreThresholds(cells_AUC,  assign=TRUE) 
    cells_assignment
    
  })  
  output$geneset_making<-renderPrint({
    if(input$cell_geneset==0){
      return()
    }
    feta_name=c("Chemosensory - Signature","Astrocyte68","Antigen presenting cell (RPS high)41","Adrenal gland inflammatory cell102","AT2 cell30","b_cell","b_cell_plasmocyte","Basal_cell","CB CD34+23","Chondrocyte99","Dendritic_cell ","Endothelial_cell  ","Endothelial cell (endothelial to mesenchymal transition)66","Endothelial cell (APC)8","Enterocyte_progenitor","Enterocyte","Epithelial cell (intermediated)60","Epithelial_cell","Erythroid_cell ","Erythroid progenitor cell (RP high)12","Fasciculata_cell ","fetal B cells","fetal Basophil_Mast","fetal CD1C+ DCs","fetal CEPs","fetal CLEC9A+ DCs","fetal Collecting duct lineage","fetal Connecting tubule lineage","fetal Distal tubule lineage","fetal EBMPs","fetal EEPs","fetal Enteroendocrine cells_intestine_pancreas","fetal Enteroendocrine cells_stomach","fetal Erythroblast","fetal ETDs","fetal HSCs","fetal HSPCs","fetal IL1B+ Microglia","fetal ILC 3","fetal Islet beta cells","fetal Islet delta cells","fetal LEC","fetal Loop of Henle lineage","fetal Macrophages","fetal Meg progenitors","fetal Meg","fetal Megakaryoblasts","fetal Microglia","fetal Nephron progenitor cell","fetal NK cells","fetal pDCs","fetal Perivascular macrophages","fetal Phagocytic macrophages","fetal Plasma cells","fetal Podocyte lineage","fetal Proximal tubule lineage","fetal PTPRC+ Microglia","fetal Pulmonary neuroendocrine cells","fetal Renal vesicle","fetal S100A9+ DCs","fetal T cells","fetal TMEM119+ Microglia","fetal TRAF1+ APCs","fetal Ureteric tip","fetal Ureteric trunk","fetalAntigen-presenting macrophages","fetalEndocardium","VEC-adrenal","VEC-brain","VEC-kidney","VEC-liver","VEC-lung","VEC-placenta","VEC-spleen","Fetal acinar cell71","Fetal chondrocyte43","Fetal endocrine cell88","Fetal enterocyte 15","Fetal epithelial progenitor1","Fetal fibroblast17","Fetal skeletal muscle cell64","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Fibroblast","Gastric chief cell50","Gastric endocrine cell84","Goblet_cell ","Hepatocyte_Endodermal cell16","hESC87","Immature sertoli cell (Pre-Sertoli cell)92","Intercalated cell56","Intermediated cell97","Kidney intercalated cell101","Loop of Henle25","M2 Macrophage47","Neutrophil (RPS high)40","Neutrophil","Macrophage","Mast cell86","Myeloid cell93","Oligodendrocyte28","Pancreas exocrine cell44","Primordial germ cell62","Proximal tubule progenitor83","Proliferating T cell52","Sinusoidal endothelial cell61","Stromal_cell","Smooth_muscle_cell","Stratified_epithelial_cell","T_cell","Thyroid follicular cell32","Ventricle cardiomyocyte74","Ureteric bud cell77")
    abc=list()
    for( i in feta_name)
    { 
      fetal_name=read.csv(paste("./",i,".tsv", sep =""))
      fetal_name<-unique(fetal_name[,1])
      fetal_list=as.vector(fetal_name)
      
      abc[[i]]=GeneSet(fetal_list,setName=i)
    }
    gensets<-GeneSetCollection(abc)
  })
  
  output$cell_rank<-renderPlot({
    if(input$action1==0){
      return()
    }
   
    
    au_mat<-pb@assays$RNA@counts
    au_mat<-as.matrix(au_mat)
    par("mar")
   
  #  geneSet<-GeneSetCollection(list(gs1,gs2,gs3,gs4,gs5,gs6,gs7,gs8,gs9,gs61,gs10,gs11,gs62,gs12,gs13,gs14,gs15,gs16,gs17,gs18,gs19,gs20,gs60,gs21,gs22,gs23,gs24,gs25,gs26,gs27,gs28,gs29,gs30,gs31,gs32,gs33,gs34,gs35,gs36,gs37,gs38,gs39,gs40,gs41,gs42,gs43,gs44,gs45,gs46,gs47,gs48,gs49,gs50,gs51,gs52,gs53,gs54,gs55,gs56,gs57))
    
    
    cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
    cells_AUC <- AUCell_calcAUC(gensets, cells_rankings) 
    
    
  })
  output$cell_auc<-renderPrint({
    if(input$cell_Auc==0){
      return()
    } 
    
    au_mat<-pb@assays$RNA@counts
    au_mat<-as.matrix(au_mat)
    #par("mar")
    
   # geneSet<-GeneSetCollection(list(gs1,gs2,gs3,gs4,gs5,gs6,gs7,gs8,gs9,gs61,gs10,gs11,gs62,gs12,gs13,gs14,gs15,gs16,gs17,gs18,gs19,gs20,gs60,gs21,gs22,gs23,gs24,gs25,gs26,gs27,gs28,gs29,gs30,gs31,gs32,gs33,gs34,gs35,gs36,gs37,gs38,gs39,gs40,gs41,gs42,gs43,gs44,gs45,gs46,gs47,gs48,gs49,gs50,gs51,gs52,gs53,gs54,gs55,gs56,gs57))
    
    
    cells_rankings <- AUCell_buildRankings(au_mat, nCores=1)
    
    cells_AUC <- AUCell_calcAUC(gensets, cells_rankings) 
    cells_AUC
   
    
})
  output$hmap<-renderPlot({
    if(input$au4==0){
      return()
    }
    au_mat<-pb@assays$RNA@counts
    au_mat<-as.matrix(au_mat)
    par("mar")
    
   # geneSet<-GeneSetCollection(list(gs1,gs2,gs3,gs4,gs5,gs6,gs7,gs8,gs9,gs61,gs10,gs11,gs62,gs12,gs13,gs14,gs15,gs16,gs17,gs18,gs19,gs20,gs60,gs21,gs22,gs23,gs24,gs25,gs26,gs27,gs28,gs29,gs30,gs31,gs32,gs33,gs34,gs35,gs36,gs37,gs38,gs39,gs40,gs41,gs42,gs43,gs44,gs45,gs46,gs47,gs48,gs49,gs50,gs51,gs52,gs53,gs54,gs55,gs56,gs57))
    
    
    cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
    cells_rankings
    save(cells_rankings, file="cells_rankings.RData")
    cells_AUC <- AUCell_calcAUC(gensets, cells_rankings)
    save(cells_AUC, file="cells_AUC.RData")    
    set.seed(123)
    par(mfrow=c(1,1)) 
    cells_assignment <- AUCell_exploreThresholds(cells_AUC,  assign=TRUE) 
     cellsAssigned <- lapply(cells_assignment, function(x) x$assignment)
    assignmentTable <- reshape2::melt(cellsAssigned, value.name="cell")
    colnames(assignmentTable)[2] <- "gensets"
    head(assignmentTable)
    assignmentMat <- table(assignmentTable[,"gensets"], assignmentTable[,"cell"])
    assignmentMat[,1:2]
    set.seed(123)
    miniAssigMat <- assignmentMat[,sample(1:ncol(assignmentMat),100)]
    library(NMF)
    aheatmap(miniAssigMat, scale="none", color="pink")
  })
  
  
output$tsne<-renderPlot({
  if(input$action1==0){
    return()
  } 
  pb <<- RunUMAP(pb, dims = 1:10)
  DimPlot(pb, reduction = "umap",pt.size = 3)
  
})

  output$c_tsne<-renderPlot({
    if(input$action1==0){
      return()
    }
    au_mat<-pb@assays$RNA@counts
    au_mat<-as.matrix(au_mat)
    
    
    
    cells_rankings <- AUCell_buildRankings(au_mat, nCores=1, plotStats=TRUE)
    cells_rankings
    save(cells_rankings, file="cells_rankings.RData")
    cells_AUC <- AUCell_calcAUC(abc[[input$aucell]], cells_rankings) 
    cells_AUC
    set.seed(123)
    par(mfrow=c(1,1)) 
    cells_assignment <- AUCell_exploreThresholds(cells_AUC, plotHist=TRUE, assign=TRUE) 
    sumByGene <- apply(au_mat, 1, sum)
    exprMatSubset <- au_mat[which(sumByGene>0),]
    logMatrix <- log2(exprMatSubset+1)
   
    umap_mat<-pb[["umap"]]@cell.embeddings
    umap_mat<-as.data.frame(umap_mat)
    selectedThresholds <- getThresholdSelected(cells_assignment)
   
    for(geneSetName in names(selectedThresholds))
    {
      nBreaks <- 5 # Number of levels in the color palettes
      # Color palette for the cells that do not pass the threshold
      colorPal_Neg <- grDevices::colorRampPalette(c("black","blue", "skyblue"))(nBreaks)
      # Color palette for the cells that pass the threshold
      colorPal_Pos <- grDevices::colorRampPalette(c("pink", "magenta", "red"))(nBreaks)
      
      # Split cells according to their AUC value for the gene set
      passThreshold <- getAUC(cells_AUC)[geneSetName,] >  selectedThresholds[geneSetName]
      if(sum(passThreshold) >0 )
      {
        aucSplit <- split(getAUC(cells_AUC)[geneSetName,], passThreshold)
        
        # Assign cell color
        cellColor <- c(setNames(colorPal_Neg[cut(aucSplit[[1]], breaks=nBreaks)], names(aucSplit[[1]])), 
                       setNames(colorPal_Pos[cut(aucSplit[[2]], breaks=nBreaks)], names(aucSplit[[2]])))
        
        # Plot
        plot(umap_mat, main=geneSetName,
             sub="Pink/red cells pass the threshold",
             col=cellColor[rownames(umap_mat)], pch=16) 
      }
    }
   
  })
  
  output$cell_umap <- downloadHandler(
    filename = function(){'umap.pdf'},
    content = function(file){
      pdf(file)
      print(      
        # Plot

            plot(umap_mat, main=geneSetName,
                 sub="Pink/red cells pass the threshold",
                 col=cellColor[rownames(umap_mat)], pch=16) 
        
        
        
      )       
      dev.off()
    })
  output$cell_geneno <- downloadHandler(
    filename = function(){'combined.pdf'},
    content = function(file){
      pdf(file)
      print(      
         plot(umap_mat, axes=FALSE, xlab="", ylab="",
       sub="Number of detected genes",
       col=cellColorNgenes[rownames(umap_mat)], pch=16)
  # Plot
      )
           
      dev.off()
    })
 # output$hpa_set<-renderUI({
  ##  if(input$hpa==0){
     # return()
  #  }
    
   # selectInput("hpa11","Select Tissue",choices=c("Lymph Node",	"Tonsil",	"Appendix",	"Spleen",	"Bone Marrow"	,"Esophagus"	,"Skin",	"Colon",	"Rectum",	"Duodenum",	"Small Intestine",	"Stomach",	"Adipose Tissue",	"Lung",	"Placenta",	"Gallbladder",	"Urinary Bladder","Endometrium",	"Smooth Muscle","Fallopian Tube",	"Thyroid Gland",	"Ovary","Prostate",	"Kidney",	"Adrenal Gland",	"Brain",	"Salivary Gland","Pancreas",	"Skeletal Muscle",	"Liver",	"Heart Muscle"))
 # })
  
  output$hpa_stouffer<-renderPlot({
    if(input$action1==0){
      return()
    }
    else if(input$hpa11=="Lymph Node")
      
    {
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - lymph.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if(input$hpa11=="Tonsil"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - tonsil.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    
    else if(input$hpa11=="Appendix") {
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - appendix.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Spleen"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
        media_genes <- read.csv("hpa - spleen.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
        else if (input$hpa11=="Bone Marrow"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
        media_genes <- read.csv("hpa - bonemarrow.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if (input$hpa11=="Esophagus"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - esophagus.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Skin"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - skin.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Colon"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - colon.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Rectum"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - rectum.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Duodenum"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - duodenum.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Small Intestine"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - smallintestine.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Lung"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - lung.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Stomach"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - stomach.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Placenta"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - placenta.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Adipose Tissue"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - adipose tissue.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Gallbladder"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - gallbladder.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Urinary Bladder"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - urinarybladder.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Endometrium"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - endometrium.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Smooth Muscle"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - smoothmuscle.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}  
    else if (input$hpa11=="Fallopian Tube"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - fallopiantube.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Thyroid Gland"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - thyroidgland.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    else if (input$hpa11=="Spleen"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - spleen.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Ovary"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - ovary.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Prostate"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - prostate.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Kidney"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - kidney.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Adrenal Gland"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - adrenalgland.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Brain"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - brain.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Salivary Gland"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - salivarygland.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Pancreas"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - pancreas.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Skeletal Muscle"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - skeletalmuscle.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Liver"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - liver.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else {
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa- heartmuscle.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
  })
  
  output$hpa_set2<-renderTable({
    if(input$action1==0){
      return()
    }
    else if(input$hpa11=="Lymph Node")
      
    {
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - lymph.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if(input$hpa11=="Tonsil"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - tonsil.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    
    else if(input$hpa11=="Appendix") {
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - appendix.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Spleen"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - spleen.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Bone Marrow"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - bonemarrow.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    
    else if (input$hpa11=="Esophagus"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - esophagus.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Skin"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - skin.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Colon"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - colon.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Rectum"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - rectum.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Duodenum"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - duodenum.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Small Intestine"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - smallintestine.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Lung"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - lung.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Stomach"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - stomach.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Placenta"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - placenta.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Adipose Tissue"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - adipose tissue.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Gallbladder"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - gallbladder.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Urinary Bladder"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - urinarybladder.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Endometrium"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - endometrium.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Smooth Muscle"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - smoothmuscle.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}  
    else if (input$hpa11=="Fallopian Tube"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - fallopiantube.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Thyroid Gland"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - thyroidgland.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    else if (input$hpa11=="Spleen"){
            
            matric<-pb@assays$RNA@counts
            
            cells_sum <- Matrix::colSums(matric>=3)
            
            matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
            
            matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
            
            cells_sum<-Matrix::rowSums(t(matric))
            
            matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
            r_name<-rownames(matric)
            media_genes <- read.csv("hpa - spleen.csv")
            same_genes<-intersect(r_name,media_genes[,1])
            if(length(same_genes)==1)
              message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
            if(length(same_genes)==0)
              message="no common gene "
            if(length(same_genes)>1){
              matric= matric[same_genes,]####### didnot work why?
              
              #Further insuring have cells with atleast 10% expressed genes
              matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
              print(dim(matric))
              seurat_RNA_matric<-matric
              print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
              print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
              print(dim(seurat_RNA_matric))
              ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
              seurat_RNA_matric[seurat_RNA_matric==0]=1
              seurat_RNA_matric<-log2(seurat_RNA_matric)
              seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
              ###Stouffer score and then boxplot for each cell type
              stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
              print(sum(is.na(stouffer_score)))
              cell_types<-Idents(pb)[names(stouffer_score)]
              unique_cell_types=unique(cell_types)
              Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
              ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
                geom_boxplot(position=position_dodge(.2)) +
                geom_jitter(shape=16, position=position_jitter(.1)) +
                theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Ovary"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - ovary.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Prostate"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - prostate.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Kidney"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - kidney.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Adrenal Gland"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - adrenalgland.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Brain"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - brain.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
    else if (input$hpa11=="Salivary Gland"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - salivarygland.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Pancreas"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - pancreas.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else if (input$hpa11=="Skeletal Muscle"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - skeletalmuscle.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    else if (input$hpa11=="Liver"){
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa - liver.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
    else {
      
      matric<-pb@assays$RNA@counts
      
      cells_sum <- Matrix::colSums(matric>=3)
      
      matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
      
      matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
      
      cells_sum<-Matrix::rowSums(t(matric))
      
      matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
      r_name<-rownames(matric)
      media_genes <- read.csv("hpa- heartmuscle.csv")
      same_genes<-intersect(r_name,media_genes[,1])
      if(length(same_genes)==1)
        message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
      if(length(same_genes)==0)
        message="no common gene "
      if(length(same_genes)>1){
        matric= matric[same_genes,]####### didnot work why?
        
        #Further insuring have cells with atleast 10% expressed genes
        matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
        print(dim(matric))
        seurat_RNA_matric<-matric
        print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
        print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
        print(dim(seurat_RNA_matric))
        ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
        seurat_RNA_matric[seurat_RNA_matric==0]=1
        seurat_RNA_matric<-log2(seurat_RNA_matric)
        seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
        ###Stouffer score and then boxplot for each cell type
        stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
        print(sum(is.na(stouffer_score)))
        cell_types<-Idents(pb)[names(stouffer_score)]
        unique_cell_types=unique(cell_types)
        Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
        ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
          geom_boxplot(position=position_dodge(.2)) +
          geom_jitter(shape=16, position=position_jitter(.1)) +
          theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
    p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
    colnames(p_val_stouffer_score)=unique_cell_types
    rownames(p_val_stouffer_score)=unique_cell_types
    for (i in 1:dim(p_val_stouffer_score)[1])
    {
      for (j in 1:dim(p_val_stouffer_score)[2])
      {
        print(paste(i,"_",j,sep=""))
        a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
        b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
        p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
      }
    }
    format(round(p_val_stouffer_score, 4))
    
  })
  
# output$gtex<-renderUI({
 #  if(input$gtex1==0){
  #   return()
 #  }
   
#   selectInput("gtex11","Select Tissue",choices=c("Adrenal Gland","Adipose Tissue","Brain","Colon","Esophagus","Fallopian Tube","Bladder","Heart", "Muscle","Kidney","Liver","Lung","Ovary","Prostate","Salivary Gland","Skin","Small Intestine","Spleen","Stomach","Thyroid","Breast","Nerve","Blood Vessel","Uterus","Cervix/Uterine","Pituitary","Vagina","Pancreas"))
 #})
  
 output$gtex_stouffer<-renderPlot({
   if(input$action1==0){
     return()
   }
   else if(input$gtex11=="Adrenal Gland"){
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("adrenalgland_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if(input$gtex11=="Adipose Tissue")
     
   {
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("adiposetissue_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   
   else if(input$gtex11=="Brain"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("brain_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   
   
   else if(input$gtex11=="Colon") {
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("colon_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Esophagus"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("esophagus_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Fallopian Tube"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("fallopiantube_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Bladder"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("bladder_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Heart"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("heart_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Muscle"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("muscle_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Kidney"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("kideny_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Liver"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("liver_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Lung"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("lung_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Ovary"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("ovary_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Prostate"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("prostate_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Salivary Gland"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("salivarygland_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Skin"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("skin_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Small Intestine"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("smallintestine_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Spleen"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("spleen_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Stomach"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("stomach_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Thyroid"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("thyroid_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Breast"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("breast_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Nerve"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("nerve_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Blood Vessel"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("blood_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Uterus"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("uterus_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Cervix/Uterine"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("cervix_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else if (input$gtex11=="Pituitary"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("pituitary_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   
   else if (input$gtex11=="Vagina"){
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("vagina_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
   else {
     
     matric<-pb@assays$RNA@counts
     
     cells_sum <- Matrix::colSums(matric>=3)
     
     matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
     
     matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
     
     cells_sum<-Matrix::rowSums(t(matric))
     
     matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
     r_name<-rownames(matric)
     media_genes <- read.csv("Pancreas_gtex.csv")
     same_genes<-intersect(r_name,media_genes[,1])
     if(length(same_genes)==1)
       message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
     if(length(same_genes)==0)
       message="no common gene "
     if(length(same_genes)>1){
       matric= matric[same_genes,]####### didnot work why?
       
       #Further insuring have cells with atleast 10% expressed genes
       matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
       print(dim(matric))
       seurat_RNA_matric<-matric
       print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
       print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
       print(dim(seurat_RNA_matric))
       ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
       seurat_RNA_matric[seurat_RNA_matric==0]=1
       seurat_RNA_matric<-log2(seurat_RNA_matric)
       seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
       ###Stouffer score and then boxplot for each cell type
       stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
       print(sum(is.na(stouffer_score)))
       cell_types<-Idents(pb)[names(stouffer_score)]
       unique_cell_types=unique(cell_types)
       Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
       ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
         geom_boxplot(position=position_dodge(.2)) +
         geom_jitter(shape=16, position=position_jitter(.1)) +
         theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}

   
 })  
 
output$gtex_pstouffer<-renderTable({
  if(input$action1==0){
    return()
  }
  else if(input$gtex11=="Adrenal Gland"){
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adrenalgland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$gtex11=="Adipose Tissue")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adiposetissue_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$gtex11=="Brain"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("brain_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$gtex11=="Colon") {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("colon_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Esophagus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("esophagus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Fallopian Tube"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("fallopiantube_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Bladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("bladder_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Heart"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("heart_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("muscle_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Kidney"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("kideny_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Liver"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("liver_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Lung"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("lung_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Ovary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("ovary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Prostate"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("prostate_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Salivary Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("salivarygland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Skin"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("skin_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Small Intestine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("smallintestine_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Spleen"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("spleen_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Stomach"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("stomach_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Thyroid"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("thyroid_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Breast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("breast_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Nerve"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("nerve_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Blood Vessel"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("blood_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Uterus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("uterus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Cervix/Uterine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("cervix_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Pituitary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("pituitary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if (input$gtex11=="Vagina"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("vagina_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Pancreas_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
  colnames(p_val_stouffer_score)=unique_cell_types
  rownames(p_val_stouffer_score)=unique_cell_types
  for (i in 1:dim(p_val_stouffer_score)[1])
  {
    for (j in 1:dim(p_val_stouffer_score)[2])
    {
      print(paste(i,"_",j,sep=""))
      a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
      b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
      p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
    }
  }
  format(round(p_val_stouffer_score, 4))
  
  
  
})

output$ex_cellrank<-renderPrint({
  if(input$action1==0){
    return()
  }
  
  
paste("The first step to calculate the enrichment of a signature is to create the rankings. These rankings are only an intermediate step to calculate the AUC.
For each cell, the genes are ranked from highest to lowest value. The genes with the same expression value are shuffled. Therefore, genes with the expression '0' are randomly sorted at the end of the ranking.
")
})
output$ex_AUC<-renderPrint({
  if(input$action1==0){
    return()
  }
paste("To determine whether the gene set is enriched at the top of the gene-ranking for each cell, AUCell uses the Area Under the Curve (AUC) of the recovery curve. In order to calculate the AUC, by default, only the top 5% of the genes in the ranking are used (i.e. checks whether the genes in the gene-set or signature are within the top 5%). This allows faster execution on bigger datasets and reduces the effect of the noise at the bottom of the ranking (e.g. where many genes might be tied at 0 counts). ")
    })
  output$ex_assign<-renderPrint({
    if(input$action1==0){
      return()
    }
paste("The AUC represents the proportion of expressed genes in the signature and their relative expression value compared to the other genes within the cell. We can use this property to explore the population of cells that are present in the dataset according to the expression of the gene-set.
To ease the selection of an assignment threshold, this function adjusts the AUCs of each gene-set to several distributions and calculates possible thresholds:
minimumDens (plot in Blue): Inflection point of the density curve. This is usually a good option for the ideal situation with bimodal distributions. To avoid false positives, by default this threshold will not be chosen if the second distribution is higher (i.e. the majority of cells have the gene-set active).

L_k2 (plot in Red): Left distribution, after adjusting the AUC to a mixture of two distributions. The threshold is set to the right (prob: 1-(thrP/nCells)). 

R_k3 (plot in Pink): Right distribution, after adjusting the AUC to a mixture of three distributions. The threshold is set to the left (prob: thrP).

Global_k1 (plot in Grey): global distribution (i.e. mean and standard deviations of all cells). The threshold is set to the right (prob: 1-(thrP/nCells)).
")
      })
output$ex_heat<-renderPrint({
  if(input$action1==0){
    return()
  }
paste("Extract these cells for all the gene-sets and transform it into a table of 0 and 1, where 1 represents the presence of that cell in a particular geneset and the heatmap is plotted based on that.")
  })
output$ex_umappp<-renderPrint({
  if(input$action1==0){
    return()
  }
paste("UMAP is colored based on the AUC scores. To highlight the cluster of cells that are more likely of the cell type according to the signatures, we will split the cells into the cells that pass the assignment threshold (colored in shades of pink-red), and the cells that don't (colored in black-blue)")
  })
output$ex_uma<-renderPrint({
  if(input$action1==0){
    return()
  }  
  paste("  The goal of these algorithms is to learn the underlying manifold of the data in order to place similar cells together in low-dimensional space. Cells within the graph-based clusters determined above should co-localize on these dimension reduction plots. ")   
})
output$ex_stouffer<-renderPrint({
  if(input$action1==0){
    return()
  }  
paste("Stouffer's Z, a method closely related to Fisher's method, is based on Z-scores rather than p-values, allowing incorporation of study weights. It is named after the sociologist, Samuel A. Stouffer. ")
  })
output$n_gene<-renderPlot({
  if(input$ngene==0){
    return()
  } 
 
   au_mat<-pb@assays$RNA@counts
  au_mat<-as.matrix(au_mat)
  
  pb$celltype<-Idents(pb)
  umap_mat<-pb[["umap"]]@cell.embeddings
  umap_mat<-as.data.frame(umap_mat)
  
  ###########################
  
  nGenesPerCell <- apply(au_mat, 2, function(x) sum(x>0))
  colorPal <- grDevices::colorRampPalette(c("darkgreen", "yellow","red"))
  cellColorNgenes <- setNames(adjustcolor(colorPal(10), alpha.f=.8)[as.numeric(cut(nGenesPerCell,breaks=10, right=FALSE,include.lowest=TRUE))], names(nGenesPerCell))
  plot(umap_mat, axes=FALSE, xlab="", ylab="",
       main="Number of Detected Genes",
       col=cellColorNgenes[rownames(umap_mat)], pch=16)
  
  
  
  
})
output$hpa_grn_1<- renderSimpleNetwork({
  if(input$action1==0){
    return()
  } 

  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  
  tissue_name=c("hpa - adipose tissue","hpa - appendix","hpa - bonemarrow","hpa - colon","hpa - duodenum","hpa - endometrium","hpa - esophagus","hpa - gallbladder","hpa - lung","hpa - lymph","hpa - placenta","hpa - rectum","hpa - skin","hpa - smallintestine","hpa - spleen","hpa - stomach","hpa - tonsil","hpa - urinarybladder","hpa - adrenalgland","hpa - brain","hpa - fallopiantube","hpa - kidney","hpa - liver","hpa - ovary","hpa - pancreas","hpa - prostate","hpa - salivarygland","hpa - skeletalmuscle","hpa - skeletalmuscle","hpa - smoothmuscle","hpa - thyroidgland")
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
    
    #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  
  
  #P_fold__df<-subset(P_fold__df,P_fold__df$logPvalue!= 0.000000e+00 && P_fold__df$adjusatedPvalue!=0.000000e+00)
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "HPA Type"
  
  final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)
  
  data_1<-data_frame(
    from=final_hpa_grn$Target,
    to=final_hpa_grn$`HPA Type`
  )
  
 
 # p <- simpleNetwork(data_1, height="100px", width="100px")
  p<-simpleNetwork(data_1, height="200px", width="300px",linkDistance = 200, charge = -50, fontSize = 10, fontFamily = "serif",
                   linkColour = "#FFB6C1", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)
  
  #par(mfrow=c(1,1)) 
  
  
  #plotm <- as.matrix(final_hpa_grn)
  # plotg <- graph_from_edgelist(rbind(plotm[,1:2],plotm[,2:3]), directed = T)
  # l <- layout_with_sugiyama(plotg, ceiling(match(V(plotg)$name, plotm)/nrow(plotm)),vgap = 20,hgap = 20)
  #nvv_col <- brewer.pal(9, "Oranges")[6]
  # hello<- plot(plotg, layout=-l$layout[,2:1],vertex.size = 30, vertex.color = "PINK", 
  #    vertex.frame.color = "RED", vertex.label.color = "black", vertex.label.dist=1, vertex.label.cex = 1.5,main="Gene Regulatory Network")
  
  
  
})
output$hpa_grn<- renderSimpleNetwork({
  if(input$action1==0){
    return()
  } 
  
  
  
  
  
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  
  tissue_name=c("hpa - adipose tissue","hpa - appendix","hpa - bonemarrow","hpa - colon","hpa - duodenum","hpa - endometrium","hpa - esophagus","hpa - gallbladder","hpa - lung","hpa - lymph","hpa - placenta","hpa - rectum","hpa - skin","hpa - smallintestine","hpa - spleen","hpa - stomach","hpa - tonsil","hpa - urinarybladder","hpa - adrenalgland","hpa - brain","hpa - fallopiantube","hpa - kidney","hpa - liver","hpa - ovary","hpa - pancreas","hpa - prostate","hpa - salivarygland","hpa - skeletalmuscle","hpa - skeletalmuscle","hpa - smoothmuscle","hpa - thyroidgland")
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
  
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
     
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
    
      #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  
  
  #P_fold__df<-subset(P_fold__df,P_fold__df$logPvalue!= 0.000000e+00 && P_fold__df$adjusatedPvalue!=0.000000e+00)
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "HPA Type"
  
  final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)
 
data_1<-data_frame(
  from=final_hpa_grn$Target,
  to=final_hpa_grn$`HPA Type`,
)

data_2<-data_frame(
  from=final_hpa_grn$`Transcription factors`,
  to=final_hpa_grn$Target,
)
#p <- simpleNetwork(data_1, height="100px", width="100px")
q<-simpleNetwork(data_2, height="200px", width="300px",linkDistance = 200, charge = -30, fontSize = 10, fontFamily = "serif",
                 linkColour = "#90EE90", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)

  #par(mfrow=c(1,1)) 
  
  
  #plotm <- as.matrix(final_hpa_grn)
 # plotg <- graph_from_edgelist(rbind(plotm[,1:2],plotm[,2:3]), directed = T)
 # l <- layout_with_sugiyama(plotg, ceiling(match(V(plotg)$name, plotm)/nrow(plotm)),vgap = 20,hgap = 20)
  #nvv_col <- brewer.pal(9, "Oranges")[6]
 # hello<- plot(plotg, layout=-l$layout[,2:1],vertex.size = 30, vertex.color = "PINK", 
           #    vertex.frame.color = "RED", vertex.label.color = "black", vertex.label.dist=1, vertex.label.cex = 1.5,main="Gene Regulatory Network")
  
  
  
})
output$hpa_grn_table<-renderTable({
  if(input$action1==0){
    return()
  } 
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  
  tissue_name=c("hpa - adipose tissue","hpa - appendix","hpa - bonemarrow","hpa - colon","hpa - duodenum","hpa - endometrium","hpa - esophagus","hpa - gallbladder","hpa - lung","hpa - lymph","hpa - placenta","hpa - rectum","hpa - skin","hpa - smallintestine","hpa - spleen","hpa - stomach","hpa - tonsil","hpa - urinarybladder","hpa - adrenalgland","hpa - brain","hpa - fallopiantube","hpa - kidney","hpa - liver","hpa - ovary","hpa - pancreas","hpa - prostate","hpa - salivarygland","hpa - skeletalmuscle","hpa - skeletalmuscle","hpa - smoothmuscle","hpa - thyroidgland")
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
 
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
}
  }
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "HPA Datset Tissues"
  final_hpa_grn 
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
 
})
output$gtex_grn_1<-renderSimpleNetwork({
  
  if(input$action1==0){
    return()
  } 
  
  
  tissue_name=c("adrenalgland_gtex","adiposetissue_gtex","brain_gtex","colon_gtex","esophagus_gtex","fallopiantube_gtex","bladder_gtex","heart_gtex","muscle_gtex","kideny - kideny_gtex","liver_gtex","lung_gtex","ovary_gtex","prostate_gtex","salivarygland_gtex","skin_gtex","smallintestine_gtex","spleen_gtex","stomach_gtex","thyroid_gtex","breast_gtex","nerve_gtex","blood_gtex","uterus_gtex","cervix_gtex","pituitary_gtex","vagina_gtex","Pancreas_gtex")
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_gtex_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V1")] <- "Target"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V2")] <- "GTEx Tissue"
  
  final_gtex_grn$mor <- as.factor(final_gtex_grn$mor)
  
  final_gtex_grn
  
  
  data_1<-data_frame(
    from=final_gtex_grn$`Transcription factors`,
    to=final_gtex_grn$Target
  )
  
  
  # p <- simpleNetwork(data_1, height="100px", width="100px")
  p<-simpleNetwork(data_1, height="200px", width="300px",linkDistance = 200, charge = -50, fontSize = 10, fontFamily = "serif",
                   linkColour = "#90EE90", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)
  
  
})

output$gtex_grn<-renderSimpleNetwork({
  
  if(input$action1==0){
    return()
  } 
  
  
  tissue_name=c("adrenalgland_gtex","adiposetissue_gtex","brain_gtex","colon_gtex","esophagus_gtex","fallopiantube_gtex","bladder_gtex","heart_gtex","muscle_gtex","kideny - kideny_gtex","liver_gtex","lung_gtex","ovary_gtex","prostate_gtex","salivarygland_gtex","skin_gtex","smallintestine_gtex","spleen_gtex","stomach_gtex","thyroid_gtex","breast_gtex","nerve_gtex","blood_gtex","uterus_gtex","cervix_gtex","pituitary_gtex","vagina_gtex","Pancreas_gtex")
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
      }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_gtex_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V1")] <- "Target"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V2")] <- "GTEx Tissue"
  
  final_gtex_grn$mor <- as.factor(final_gtex_grn$mor)
  
  final_gtex_grn

  
  data_1<-data_frame(
    from=final_gtex_grn$Target,
    to=final_gtex_grn$`GTEx Tissue`
  )
  
  
  # p <- simpleNetwork(data_1, height="100px", width="100px")
  p<-simpleNetwork(data_1, height="200px", width="300px",linkDistance = 200, charge = -50, fontSize = 10, fontFamily = "serif",
                   linkColour = "	#FFB6C1", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)
  
  
})
output$gtex_grn_table<-renderTable({
  
  if(input$action1==0){
    return()
  } 
  
  
  tissue_name=c("adrenalgland_gtex","adiposetissue_gtex","brain_gtex","colon_gtex","esophagus_gtex","fallopiantube_gtex","bladder_gtex","heart_gtex","muscle_gtex","kideny - kideny_gtex","liver_gtex","lung_gtex","ovary_gtex","prostate_gtex","salivarygland_gtex","skin_gtex","smallintestine_gtex","spleen_gtex","stomach_gtex","thyroid_gtex","breast_gtex","nerve_gtex","blood_gtex","uterus_gtex","cervix_gtex","pituitary_gtex","vagina_gtex","Pancreas_gtex")
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
    }
  }
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_gtex_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V1")] <- "Target"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_gtex_grn)[which(names(final_gtex_grn) == "V2")] <- "HPA Tissue "
  
  final_gtex_grn$mor <- as.factor(final_gtex_grn$mor)
  
  final_gtex_grn
  
})

output$cell_grn_1<-renderSimpleNetwork({
  if(input$action1==0){
    return()
  } 
  
  ###########
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Embryonic_stem_cell","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  
  
  
  
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_cell_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V1")] <- "Target"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V2")] <- "Cell Enrich Signature"
  
  final_cell_grn$mor <- as.factor(final_cell_grn$mor)
  
  
  data_2<-data_frame(
    from=final_cell_grn$`Transcription factors`,
    to=final_cell_grn$Target
  )
  
  
  #p <- simpleNetwork(data_1, height="100px", width="100px")
  q<-simpleNetwork(data_2, height="200px", width="300px",linkDistance = 200, charge = -30, fontSize = 10, fontFamily = "serif",
                   linkColour = "#90EE90", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)
  
  ######
  
  
  
  
  
})






output$cell_grn<-renderSimpleNetwork({
  if(input$action1==0){
    return()
  } 
  
  ###########
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Embryonic_stem_cell","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  

  
  
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
}
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
 
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_cell_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V1")] <- "Target"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V2")] <- "Cell Enrich Signature"
  
  final_cell_grn$mor <- as.factor(final_cell_grn$mor)
  
  
  data_1<-data_frame(
    from=final_cell_grn$Target,
    to=final_cell_grn$`Cell Enrich Signature`
  )
  
  
  #p <- simpleNetwork(data_1, height="100px", width="100px")
  q<-simpleNetwork(data_1, height="200px", width="300px",linkDistance = 200, charge = -30, fontSize = 10, fontFamily = "serif",
                   linkColour = "#FFB6C1", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)
  
  ######
  
  
  

  
})



















cell_grn_d<-reactive({
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  
  
  
  
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_cell_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V1")] <- "Target"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V2")] <- "Cell Enrich Signature"
  
  final_cell_grn$mor <- as.factor(final_cell_grn$mor)
  ######################
  
})


output$cell_grn_table<-renderTable({
  if(input$action1==0){
    return()
  } 
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Embryonic_stem_cell","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  

  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
 
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
     }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "Cell Enrich Signature"
  final_hpa_grn
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
 
})



output$cell_stoudownload <- downloadHandler(
  filename = function(){'Stouffer CellEnrich.pdf'},
  content = function(file){
    pdf(file)
    print(  


stouffer_cellenrich()
    
    )
    dev.off()
  })


cell_grn_d<-reactive({
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  
  
  
  
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_cell_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V1")] <- "Target"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_cell_grn)[which(names(final_cell_grn) == "V2")] <- "Cell Enrich Signature"
  
  final_cell_grn$mor <- as.factor(final_cell_grn$mor)
  
  
  
  
  
  
  plotm <- as.matrix(final_cell_grn)
  plotg <- graph_from_edgelist(rbind(plotm[,1:2],plotm[,2:3]), directed = T)
  l <- layout_with_sugiyama(plotg, ceiling(match(V(plotg)$name, plotm)/nrow(plotm)),vgap = 20,hgap = 20)
  #nvv_col <- brewer.pal(9, "Oranges")[6]
  hello<- plot(plotg, layout=-l$layout[,2:1],vertex.size = 30, vertex.color = "PINK",
               vertex.frame.color = "RED", vertex.label.color = "black", vertex.label.dist=1, vertex.label.cex = 1.5,main="Gene Regulatory Network")
  ##############################
  
  
})



output$cell_grndownload <- downloadHandler(
  filename = function(){'GRN.pdf'},
  content = function(file){
    pdf(file)
    print(  
      
      cell_grn_d()
      
      
    )
    dev.off()
  })



hpa_grn_d<-reactive({total_tissue_gene<-scan("gene_name_cluster.txt",character())

tissue_name=c("hpa - adipose tissue","hpa - appendix","hpa - bonemarrow","hpa - colon","hpa - duodenum","hpa - endometrium","hpa - esophagus","hpa - gallbladder","hpa - lung","hpa - lymph","hpa - placenta","hpa - rectum","hpa - skin","hpa - smallintestine","hpa - spleen","hpa - stomach","hpa - tonsil","hpa - urinarybladder","hpa - adrenalgland","hpa - brain","hpa - fallopiantube","hpa - kidney","hpa - liver","hpa - ovary","hpa - pancreas","hpa - prostate","hpa - salivarygland","hpa - skeletalmuscle","hpa - skeletalmuscle","hpa - smoothmuscle","hpa - thyroidgland")
user_genes<-read.table("gene_name_cluster.txt")


# tissue_name=list("heart_human","placenta_human","placenta_mouse")


hpa_cell=c()

for( i in tissue_name)
{ 
  tissue_genes=read.csv(paste("./",i,".csv", sep =""))
  genes_list=intersect(tissue_genes[,1],user_genes[,1])
  # total_tissue<-scan("tissue_genes",character())
  #total_tissue_gene<-scan("user_genes",character())
  
  if(length(genes_list)>=1)
    
  { 
    write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
    temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
    hpa_cell=rbind(hpa_cell,temp)
    
    #f1<-(length_overlap/length_tissue_genes)
    #f2<- (tissue_genes_length/20400)
    #fold_change<-f1/f2 incorrect number of subscripts on matrix
  }
  
  #P_fold[i,]=c(p,fold_change,length(genes_list))  
}



#P_fold__df<-subset(P_fold__df,P_fold__df$logPvalue!= 0.000000e+00 && P_fold__df$adjusatedPvalue!=0.000000e+00)

hpa_common<-as.data.frame(hpa_cell)
pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold

##########************************* read dorothea_regulon file
dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")

## We obtain the regulons based on interactions with confidence level A, B and C
regulon <- dorothea_regulon_human %>%
  dplyr::filter(confidence %in% c("A","B","C")) 

## We compute Viper Scores 
pb <- run_viper(pb, regulon)

## We compute the Nearest Neighbours to perform cluster
DefaultAssay(object = pb) <- "dorothea"
pb <- ScaleData(pb)
pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)

pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")

pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                             logfc.threshold = input$log_fc , verbose = FALSE)

DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()

## We transform Viper scores, scaled by seurat, into a data frame to better 

## handling the results
viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                assay = "dorothea") %>%
  data.frame() %>%
  t()

## We create a data frame containing the cells and their clusters
CellsClusters <- data.frame(cell = names(Idents(pb)), 
                            cell_type = as.character(Idents(pb)),
                            stringsAsFactors = FALSE)


################## my addition ###################################
viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)

summarized_viper_scores <- viper_scores_clusters %>% 
  group_by(variable, cell_type) %>%
  summarise(avg = mean(value),
            std = sd(value))
## We select the 20 most variable TFs. (20*9 populations = 180)

highly_variable_tfs <- summarized_viper_scores %>%
  group_by(variable) %>%
  mutate(var = var(avg))  %>%
  ungroup() %>%
  top_n(180, var) %>%
  distinct(variable)

## We prepare the data for the plot
summarized_viper_scores_df <- summarized_viper_scores %>%
  semi_join(highly_variable_tfs, by = "variable") %>%
  dplyr::select(-std) %>%   
  spread(variable, avg) %>%
  data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 

transpose_summarise_df<-t(summarized_viper_scores_df)

merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "hpa Type"

final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)

data_1<-data_frame(
  from=final_hpa_grn$Target,
  to=final_hpa_grn$`HPA Type`,
)


# p <- simpleNetwork(data_1, height="100px", width="100px")
p<-simpleNetwork(data_1, height="200px", width="300px",linkDistance = 200, charge = -50, fontSize = 10, fontFamily = "serif",
                 linkColour = "	#FF69B4", nodeColour = "#FFFFFF", opacity = 1, zoom = T	)

#par(mfrow=c(1,1)) 





  
})

gtex_d<-reactive({
  tissue_name=c("adrenalgland_gtex","adiposetissue_gtex","brain_gtex","colon_gtex","esophagus_gtex","fallopiantube_gtex","bladder_gtex","heart_gtex","muscle_gtex","kideny - kideny_gtex","liver_gtex","lung_gtex","ovary_gtex","prostate_gtex","salivarygland_gtex","skin_gtex","smallintestine_gtex","spleen_gtex","stomach_gtex","thyroid_gtex","breast_gtex","nerve_gtex","blood_gtex","uterus_gtex","cervix_gtex","pituitary_gtex","vagina_gtex","Pancreas_gtex")
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "GTEx Tissue "
  
  final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)
  
  final_hpa_grn
  
  
})









gtex_grn_d<-reactive({
  tissue_name=c("adrenalgland_gtex","adiposetissue_gtex","brain_gtex","colon_gtex","esophagus_gtex","fallopiantube_gtex","bladder_gtex","heart_gtex","muscle_gtex","kideny - kideny_gtex","liver_gtex","lung_gtex","ovary_gtex","prostate_gtex","salivarygland_gtex","skin_gtex","smallintestine_gtex","spleen_gtex","stomach_gtex","thyroid_gtex","breast_gtex","nerve_gtex","blood_gtex","uterus_gtex","cervix_gtex","pituitary_gtex","vagina_gtex","Pancreas_gtex")
  
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "HPA Tissue "
  
  final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)
  
  final_hpa_grn
  
  
  
  plotm <- as.matrix(final_hpa_grn)
  plotg <- graph_from_edgelist(rbind(plotm[,1:2],plotm[,2:3]), directed = T)
  l <- layout_with_sugiyama(plotg, ceiling(match(V(plotg)$name, plotm)/nrow(plotm)))
  #nvv_col <- brewer.pal(9, "Oranges")[6]
  hello<- plot(plotg, layout=-l$layout[,2:1],vertex.size = 30, vertex.color = "PINK",
               vertex.frame.color = "RED", vertex.label.color = "black", vertex.label.dist=1, vertex.label.cex = 1.5,main="Gene Regulatory Network")
  
})

cell_grn_dt<-reactive({
  tissue_name=c("Devloping heart atrial cardiomyocyte","Devloping heart ventricular cardiomyocyte","Ectodermal cell differentiation","Ectodermal development","Endodermal cell differentiation","Endodermal development","Fibroblast migration","Fibroblast proliferation","Myofibroblast differentiation","Mesoderm development","Mesodermal cell differentiation","Adult astrocyte","Adult antigen presenting cell (RPS high)","Adult monocyte","Adult adrenal gland inflammatory cell","Adult AT2 cell","Adult b_cell","Adult b_cell_plasmocyte","Adult basal_cell","Adult CB CD34+23","Adult chondrocyte","Adult chemosensory ","Adult dendritic_cell ","Adult endothelial_cell","Adult endothelial cell (endothelial to mesenchymal transition)","Adult endothelial cell (APC)","Adult enterocyte_progenitor","Adult enterocyte","Adult epithelial cell (intermediated)","Adult epithelial_cell","Adult erythroid_cell ","Adult erythroid progenitor cell (RP high)","Adult fasciculata_cell ","Fetal B cells","Fetal Basophil_Mast","Fetal CD1C+ DCs","Fetal CEPs","Fetal CLEC9A+ DCs","Fetal Collecting duct lineage","Fetal Connecting tubule lineage","Fetal Distal tubule lineage","Fetal EBMPs","Fetal EEPs","Fetal Enteroendocrine cells_intestine_pancreas","Fetal Enteroendocrine cells_stomach","Fetal Erythroblast","Fetal ETDs","Fetal HSCs","Fetal HSPCs","Fetal IL1B+ Microglia","Fetal ILC 3","Fetal Islet beta cells","Fetal Islet delta cells","Fetal LEC","Fetal Loop of Henle lineage","Fetal Macrophages","Fetal Meg progenitors","Fetal Meg","Fetal Megakaryoblasts","Fetal Microglia","Fetal Nephron progenitor cell","Fetal NK cells","Fetal pDCs","Fetal Perivascular macrophages","Fetal Phagocytic macrophages","Fetal Plasma cells","Fetal Podocyte lineage","Fetal Proximal tubule lineage","Fetal PTPRC+ Microglia","Fetal Pulmonary neuroendocrine cells","Fetal Renal vesicle","Fetal S100A9+ DCs","Fetal T cells","Fetal TMEM119+ Microglia","Fetal TRAF1+ APCs","Fetal Ureteric tip","Fetal Ureteric trunk","Fetal Antigen-presenting macrophages","Fetal Endocardium","Fetal adrenal","Fetal brain","Fetal kidney","Fetal liver","Fetal lung","Fetal placenta","Fetal spleen","Fetal acinar cell","Fetal chondrocyte","Fetal endocrine cell","Fetal enterocyte","Fetal epithelial progenitor","Fetal fibroblast","Fetal skeletal muscle cell","Fetal_mesenchymal_progenitor","Fetal_neuron ","Fetal_stromal_cell","Adult fibroblast","Adult gastric chief cell","Adult gastric endocrine cell","Adult goblet_cell ","Adult hepatocyte_Endodermal cell","Adult hESC87","Adult immature sertoli cell (Pre-Sertoli cell)","Adult intercalated cell","Adult intermediated cell","Adult kidney intercalated cell","Adult loop of Henle","Adult M2 Macrophage","Adult neutrophil (RPS high)","Adult neutrophil","Adult macrophage","Adult mast cell","Adult myeloid cell","Adult oligodendrocyte","Adult pancreas exocrine cell","Adult primordial germ cell","Adult proximal tubule progenitor","Adult proliferating T cell","Adult sinusoidal endothelial cell","Adult stromal_cell","Adult smooth_muscle_cell","Adult stratified_epithelial_cell","Adult T_cell","Adult thyroid follicular cell","Adult ventricle cardiomyocyte","Adult ureteric bud cell")
  
  
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".tsv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    x<-nrow(pb)
    if(length(genes_list)>1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
    }  #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc , verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "Cell Enrich Signature"
  final_hpa_grn
  # tissue_name=list("
  
 
})
hpa_grn_dt<-reactive({
  total_tissue_gene<-scan("gene_name_cluster.txt",character())
  
  tissue_name=c("hpa - adipose tissue","hpa - appendix","hpa - bonemarrow","hpa - colon","hpa - duodenum","hpa - endometrium","hpa - esophagus","hpa - gallbladder","hpa - lung","hpa - lymph","hpa - placenta","hpa - rectum","hpa - skin","hpa - smallintestine","hpa - spleen","hpa - stomach","hpa - tonsil","hpa - urinarybladder","hpa - adrenalgland","hpa - brain","hpa - fallopiantube","hpa - kidney","hpa - liver","hpa - ovary","hpa - pancreas","hpa - prostate","hpa - salivarygland","hpa - skeletalmuscle","hpa - skeletalmuscle","hpa - smoothmuscle","hpa - thyroidgland")
  user_genes<-read.table("gene_name_cluster.txt")
  
  
  # tissue_name=list("heart_human","placenta_human","placenta_mouse")
  
  
  hpa_cell=c()
  
  for( i in tissue_name)
  { 
    tissue_genes=read.csv(paste("./",i,".csv", sep =""))
    genes_list=intersect(tissue_genes[,1],user_genes[,1])
    # total_tissue<-scan("tissue_genes",character())
    #total_tissue_gene<-scan("user_genes",character())
    
    if(length(genes_list)>=1)
      
    { 
      write.csv(genes_list,paste(i,"_common_genes.csv",sep=""))
      temp=matrix(c(genes_list,rep(i,length(genes_list))),ncol=2)
      hpa_cell=rbind(hpa_cell,temp)
      
      #f1<-(length_overlap/length_tissue_genes)
      #f2<- (tissue_genes_length/20400)
      #fold_change<-f1/f2 incorrect number of subscripts on matrix
    }
    
    #P_fold[i,]=c(p,fold_change,length(genes_list))  
  }
  
  
  
  #P_fold__df<-subset(P_fold__df,P_fold__df$logPvalue!= 0.000000e+00 && P_fold__df$adjusatedPvalue!=0.000000e+00)
  
  hpa_common<-as.data.frame(hpa_cell)
  pb.markers <<- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, logfc.threshold = input$log_fc)### have to make chnage in threshold
  
  ##########************************* read dorothea_regulon file
  dorothea_regulon_human <-read.csv("dorothea_regulon_human.csv")
  
  ## We obtain the regulons based on interactions with confidence level A, B and C
  regulon <- dorothea_regulon_human %>%
    dplyr::filter(confidence %in% c("A","B","C")) 
  
  ## We compute Viper Scores 
  pb <- run_viper(pb, regulon)
  
  ## We compute the Nearest Neighbours to perform cluster
  DefaultAssay(object = pb) <- "dorothea"
  pb <- ScaleData(pb)
  pb <- RunPCA(pb, features = rownames(pb), verbose = FALSE)
  pb <- FindNeighbors(pb, dims = 1:10, verbose = FALSE)
  pb <- FindClusters(pb, resolution = 0.5, verbose = FALSE)
  
  pb <- RunUMAP(pb, dims = 1:10, umap.method = "uwot", metric = "cosine")
  
  pb.markers <- FindAllMarkers(pb, only.pos = TRUE, min.pct = 0.25, 
                               logfc.threshold = input$log_fc, verbose = FALSE)
  
  DimPlot(pb, reduction = "umap", label = FALSE, pt.size = 0.5) + NoLegend()
  
  ## We transform Viper scores, scaled by seurat, into a data frame to better 
  
  ## handling the results
  viper_scores_df <- GetAssayData(pb, slot = "scale.data", 
                                  assay = "dorothea") %>%
    data.frame() %>%
    t()
  
  ## We create a data frame containing the cells and their clusters
  CellsClusters <- data.frame(cell = names(Idents(pb)), 
                              cell_type = as.character(Idents(pb)),
                              stringsAsFactors = FALSE)
  
  
  ################## my addition ###################################
  viper_scores_clusters <- viper_scores_df  %>% data.frame() %>% rownames_to_column("cell")
  viper_scores_clusters<- melt(viper_scores_clusters, id= "cell")
  viper_scores_clusters<- viper_scores_clusters %>% inner_join(CellsClusters)
  
  summarized_viper_scores <- viper_scores_clusters %>% 
    group_by(variable, cell_type) %>%
    summarise(avg = mean(value),
              std = sd(value))
  ## We select the 20 most variable TFs. (20*9 populations = 180)
  
  highly_variable_tfs <- summarized_viper_scores %>%
    group_by(variable) %>%
    mutate(var = var(avg))  %>%
    ungroup() %>%
    top_n(180, var) %>%
    distinct(variable)
  
  ## We prepare the data for the plot
  summarized_viper_scores_df <- summarized_viper_scores %>%
    semi_join(highly_variable_tfs, by = "variable") %>%
    dplyr::select(-std) %>%   
    spread(variable, avg) %>%
    data.frame(row.names = 1, check.names = FALSE, stringsAsFactors = FALSE) 
  
  transpose_summarise_df<-t(summarized_viper_scores_df)
  
  merged.data <- merge(transpose_summarise_df, regulon,by.x="row.names", by.y="tf")
  final_data_plot<- merged.data  %>% dplyr::select(Row.names,confidence,target,mor)
  final_hpa_grn<-merge(hpa_common,final_data_plot,by.x="V1",by.y="target")
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V1")] <- "Target"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "Row.names")] <- "Transcription factors"
  colnames(final_hpa_grn)[which(names(final_hpa_grn) == "V2")] <- "hpa Type"
  
  final_hpa_grn$mor <- as.factor(final_hpa_grn$mor)
  
  final_hpa_grn
 
  
})





output$hpa_grndownload <- downloadHandler(
  filename = function(){'GRN.pdf'},
  content = function(file){
    pdf(file)
    print(  
      hpa_grn_d()
      
      
      
    )
    dev.off()
  })
output$gtex_grndownload <- downloadHandler(
  filename = function(){'GRN.pdf'},
  content = function(file){
    pdf(file)
    print(  
      
      gtex_grn_d()
      
      
    )
    dev.off()
  })

output$report <- downloadHandler(
  # For PDF output, change this to "report.pdf"
  filename = "report.html",
  content = function(file) {
    # Copy the report file to a temporary directory before processing it, in
    # case we don't have write permissions to the current working dir (which
    # can happen when deployed).
    tempReport <- file.path(tempdir(), "report.Rmd")
    file.copy("report.Rmd", tempReport, overwrite = TRUE)
    
    # Set up parameters to pass to Rmd document
    params <- list(dataTables)
    
    # Knit the document, passing in the `params` list, and eval it in a
    # child of the global environment (this isolates the code in the document
    # from the code in this app).
    rmarkdown::render(tempReport, output_file = file,
                      params = params,
                      envir = new.env(parent = globalenv())
    )
  }
)

stouffer_hpa<-reactive({
  if(input$hpa11=="Lymph Node")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - lymph.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$hpa11=="Tonsil"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - tonsil.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$hpa11=="Appendix") {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - appendix.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Spleen"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - spleen.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Bone Marrow"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - bonemarrow.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if (input$hpa11=="Esophagus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - esophagus.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Skin"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - skin.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Colon"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - colon.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Rectum"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - rectum.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Duodenum"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - duodenum.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Small Intestine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - smallintestine.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Lung"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - lung.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Stomach"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - stomach.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Placenta"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - placenta.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Adipose Tissue"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - adipose tissue.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Gallbladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - gallbladder.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Urinary Bladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - urinarybladder.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Endometrium"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - endometrium.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Smooth Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - smoothmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}  
  else if (input$hpa11=="Fallopian Tube"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - fallopiantube.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Thyroid Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - thyroidgland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    else if (input$hpa11=="Spleen"){
          
          matric<-pb@assays$RNA@counts
          
          cells_sum <- Matrix::colSums(matric>=3)
          
          matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
          
          matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
          
          cells_sum<-Matrix::rowSums(t(matric))
          
          matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
          r_name<-rownames(matric)
          media_genes <- read.csv("hpa - spleen.csv")
          same_genes<-intersect(r_name,media_genes[,1])
          if(length(same_genes)==1)
            message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
          if(length(same_genes)==0)
            message="no common gene "
          if(length(same_genes)>1){
            matric= matric[same_genes,]####### didnot work why?
            
            #Further insuring have cells with atleast 10% expressed genes
            matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
            print(dim(matric))
            seurat_RNA_matric<-matric
            print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
            print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
            print(dim(seurat_RNA_matric))
            ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
            seurat_RNA_matric[seurat_RNA_matric==0]=1
            seurat_RNA_matric<-log2(seurat_RNA_matric)
            seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
            ###Stouffer score and then boxplot for each cell type
            stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
            print(sum(is.na(stouffer_score)))
            cell_types<-Idents(pb)[names(stouffer_score)]
            unique_cell_types=unique(cell_types)
            Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
            ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
              geom_boxplot(position=position_dodge(.2)) +
              geom_jitter(shape=16, position=position_jitter(.1)) +
              theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Ovary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - ovary.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Prostate"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - prostate.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Kidney"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - kidney.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Adrenal Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - adrenalgland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Brain"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - brain.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Salivary Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - salivarygland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Pancreas"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - pancreas.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Skeletal Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - skeletalmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Liver"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - liver.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa- heartmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
})

stouffer_gtex<-reactive({
  
  if(input$gtex11=="Adrenal Gland"){
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adrenalgland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$gtex11=="Adipose Tissue")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adiposetissue_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$gtex11=="Brain"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("brain_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$gtex11=="Colon") {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("colon_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Esophagus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("esophagus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Fallopian Tube"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("fallopiantube_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Bladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("bladder_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Heart"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("heart_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("muscle_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Kidney"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("kideny_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Liver"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("liver_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Lung"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("lung_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Ovary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("ovary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Prostate"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("prostate_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Salivary Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("salivarygland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Skin"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("skin_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Small Intestine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("smallintestine_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Spleen"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("spleen_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Stomach"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("stomach_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Thyroid"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("thyroid_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Breast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("breast_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Nerve"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("nerve_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Blood Vessel"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("blood_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Uterus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("uterus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Cervix/Uterine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("cervix_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Pituitary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("pituitary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if (input$gtex11=="Vagina"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("vagina_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Pancreas_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
})

stouffer_cellenrich<-reactive({
  if(input$st_sc =="Devloping heart atrial cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Devloping heart atrial cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Devloping heart ventricular cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Devloping heart ventricular cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Embryonic_stem_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Embryonic_stem_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Ectodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Ectodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Ectodermal development.tsv"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Ectodermal development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Endodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Endodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Endodermal development"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Endodermal development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Myofibroblast differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Myofibroblast differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Mesoderm development"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Mesoderm development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Mesodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Mesodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  
  else if(input$st_sc =="Adult chemosensory "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("final_list.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc=="Adult adrenal gland inflammatory cell")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult adrenal gland inflammatory cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal B cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal B cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$sc_st=="Fetal Basophil_Mast")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Basophil_Mast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CD1C+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CD1C+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CEPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CEPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CLEC9A+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CLEC9A+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Collecting duct lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Collecting duct lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Connecting tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Connecting tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Distal tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Distal tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal EBMPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal EBMPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal EEPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal EEPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Enteroendocrine cells_intestine_pancreas")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Enteroendocrine cells_intestine_pancreas.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Enteroendocrine cells_stomach")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Enteroendocrine cells_stomach.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Erythroblast")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Erythroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal ETDs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal ETDs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal HSCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal HSCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal HSPCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal HSPCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal IL1B+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal IL1B+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal ILC 3")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal ILC 3.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Islet beta cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Islet beta cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Islet delta cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Islet delta cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal LEC")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal LEC.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Loop of Henle lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Loop of Henle lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Meg progenitors")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Meg progenitors.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Meg")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Meg.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Megakaryoblasts")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Megakaryoblasts.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Nephron progenitor cell")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Nephron progenitor cell")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal NK cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal NK cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal pDCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal pDCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Perivascular macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Perivascular macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Phagocytic macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Phagocytic macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Plasma cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Plasma cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Podocyte lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Podocyte lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Proximal tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Proximal tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal PTPRC+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal PTPRC+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Pulmonary neuroendocrine cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Pulmonary neuroendocrine cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Renal vesicle")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Renal vesicle.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$sc_st=="Fetal S100A9+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal S100A9+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal T cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal T cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal TMEM119+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal TMEM119+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal TRAF1+ APCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal TRAF1+ APCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Ureteric tip")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Ureteric tip.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal adrenal")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal adrenal.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal brain")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal brain.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal kidney")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal kidney.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal liver")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal liver.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal lung")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal lung.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal placenta")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal placenta.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal spleen")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal spleen.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult antigen presenting cell (RPS high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult antigen presenting cell (RPS high).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult astrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult astrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult AT2_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult AT2 cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult b_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult b_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult b_cell_plasmocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult b_cell_plasmocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult basal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult basal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult CB_CD34+"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult CB CD34+23.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult chondrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult chondrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult dendritic_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult dendritic_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial cell (APC)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial cell (APC).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial cell (endothelial to mesenchymal transition)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial cell (endothelial to mesenchymal transition).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult enterocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult enterocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult enterocyte_progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult enterocyte_progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult epithelial cell (intermediated)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult epithelial cell (intermediated).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult epithelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult epithelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult erythroid cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult erythroid cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult erythroid progenitor cell (RP high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult erythroid progenitor cell (RP high).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult fasciculata_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult fasciculata_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal acinar cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal acinar cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal chondrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal chondrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Fetal endocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal endocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal enterocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal enterocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Fetal epithelial progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal epithelial progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal fibroblast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal fibroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal skeletal muscle cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal skeletal muscle cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_mesenchymal_progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_mesenchymal_progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_neuron "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_neuron .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_stromal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_stromal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult Stromal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult stromal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult fibroblast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult fibroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult gastric chief cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult gastric chief cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult gastric endocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult gastric endocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult goblet_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult goblet_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult hepatocyte_Endodermal cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult hepatocyte_Endodermal cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult hESC87"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult hESC87.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult immature sertoli cell (Pre-Sertoli cell)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult immature sertoli cell (Pre-Sertoli cell).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult intercalated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult intercalated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult intermediated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult intermediated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult kidney intercalated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult kidney intercalated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult loop of Henle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult loop of Henle.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult M2 Macrophage"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult M2 Macrophage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult macrophage"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult macrophage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult mast cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult mast cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult mesothelial cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult mesothelial cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult monocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Monocyte.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult myeloid cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult myeloid cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Neutrophil (RPS high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Neutrophil (RPS high)40.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult neutrophil"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult neutrophil.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult oligodendrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult oligodendrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult pancreas exocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult pancreas exocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult primordial germ cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult primordial germ cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult proliferating T cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult proliferating T cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult proximal tubule progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult proximal tubule progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult sinusoidal endothelial cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult sinusoidal endothelial cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult smooth_muscle_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult smooth_muscle_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult stratified_epithelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult stratified_epithelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult T_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult T_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult thyroid follicular cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult thyroid follicular cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult ureteric bud cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("dult ureteric bud cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult ventricle cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult ventricle cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
})
output$download_hpa_gt <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv( hpa_grn_dt(), file) 
    
  }) 

stouffer_tab_gtex<-reactive({
  if(input$gtex11=="Adrenal Gland"){
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adrenalgland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$gtex11=="Adipose Tissue")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("adiposetissue_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$gtex11=="Brain"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("brain_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$gtex11=="Colon") {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("colon_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Esophagus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("esophagus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Fallopian Tube"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("fallopiantube_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Bladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("bladder_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Heart"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("heart_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("muscle_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Kidney"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("kideny_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Liver"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("liver_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Lung"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("lung_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Ovary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("ovary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Prostate"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("prostate_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Salivary Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("salivarygland_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Skin"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("skin_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Small Intestine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("smallintestine_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Spleen"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("spleen_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Stomach"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("stomach_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Thyroid"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("thyroid_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Breast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("breast_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Nerve"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("nerve_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Blood Vessel"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("blood_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Uterus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("uterus_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Cervix/Uterine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("cervix_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$gtex11=="Pituitary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("pituitary_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if (input$gtex11=="Vagina"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("vagina_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Pancreas_gtex.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
  colnames(p_val_stouffer_score)=unique_cell_types
  rownames(p_val_stouffer_score)=unique_cell_types
  for (i in 1:dim(p_val_stouffer_score)[1])
  {
    for (j in 1:dim(p_val_stouffer_score)[2])
    {
      print(paste(i,"_",j,sep=""))
      a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
      b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
      p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
    }
  }
  format(round(p_val_stouffer_score, 4))
})

output$stouffer_table_gtex <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv(stouffer_tab_gtex(), file) 
    
  }) 

output$stouffer_table_hpa <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv(stouffer_tab_hpa(), file) 
    
  }) 
output$stouffer_table_cell <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv(stouffer_tab_cell(), file) 
    
  }) 

stouffer_tab_hpa<-reactive({
   if(input$hpa11=="Lymph Node")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - lymph.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$hpa11=="Tonsil"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - tonsil.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$hpa11=="Appendix") {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - appendix.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Spleen"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - spleen.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Bone Marrow"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - bonemarrow.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if (input$hpa11=="Esophagus"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - esophagus.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Skin"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - skin.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Colon"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - colon.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Rectum"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - rectum.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Duodenum"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - duodenum.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Small Intestine"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - smallintestine.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Lung"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - lung.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Stomach"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - stomach.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Placenta"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - placenta.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Adipose Tissue"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - adipose tissue.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Gallbladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - gallbladder.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Urinary Bladder"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - urinarybladder.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Endometrium"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - endometrium.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Smooth Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - smoothmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}  
  else if (input$hpa11=="Fallopian Tube"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - fallopiantube.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Thyroid Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - thyroidgland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    else if (input$hpa11=="Spleen"){
          
          matric<-pb@assays$RNA@counts
          
          cells_sum <- Matrix::colSums(matric>=3)
          
          matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
          
          matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
          
          cells_sum<-Matrix::rowSums(t(matric))
          
          matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
          r_name<-rownames(matric)
          media_genes <- read.csv("hpa - spleen.csv")
          same_genes<-intersect(r_name,media_genes[,1])
          if(length(same_genes)==1)
            message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
          if(length(same_genes)==0)
            message="no common gene "
          if(length(same_genes)>1){
            matric= matric[same_genes,]####### didnot work why?
            
            #Further insuring have cells with atleast 10% expressed genes
            matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
            print(dim(matric))
            seurat_RNA_matric<-matric
            print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
            print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
            print(dim(seurat_RNA_matric))
            ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
            seurat_RNA_matric[seurat_RNA_matric==0]=1
            seurat_RNA_matric<-log2(seurat_RNA_matric)
            seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
            ###Stouffer score and then boxplot for each cell type
            stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
            print(sum(is.na(stouffer_score)))
            cell_types<-Idents(pb)[names(stouffer_score)]
            unique_cell_types=unique(cell_types)
            Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
            ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
              geom_boxplot(position=position_dodge(.2)) +
              geom_jitter(shape=16, position=position_jitter(.1)) +
              theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Ovary"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - ovary.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Prostate"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - prostate.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Kidney"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - kidney.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Adrenal Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - adrenalgland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Brain"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - brain.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}    
  else if (input$hpa11=="Salivary Gland"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - salivarygland.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Pancreas"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - pancreas.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else if (input$hpa11=="Skeletal Muscle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - skeletalmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if (input$hpa11=="Liver"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa - liver.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}   
  else {
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("hpa- heartmuscle.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
  colnames(p_val_stouffer_score)=unique_cell_types
  rownames(p_val_stouffer_score)=unique_cell_types
  for (i in 1:dim(p_val_stouffer_score)[1])
  {
    for (j in 1:dim(p_val_stouffer_score)[2])
    {
      print(paste(i,"_",j,sep=""))
      a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
      b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
      p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
    }
  }
  format(round(p_val_stouffer_score, 4))
  
})



stouffer_tab_cell<-reactive({
  if(input$st_sc =="Devloping heart atrial cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Devloping heart atrial cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Devloping heart ventricular cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Devloping heart ventricular cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Embryonic_stem_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Embryonic_stem_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Ectodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Ectodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Ectodermal development.tsv"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Ectodermal development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Endodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Endodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Endodermal development"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Endodermal development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Myofibroblast differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Myofibroblast differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Mesoderm development"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Mesoderm development.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Mesodermal cell differentiation"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Mesodermal cell differentiation.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  
  else if(input$st_sc =="Adult chemosensory "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("final_list.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc=="Adult adrenal gland inflammatory cell")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult adrenal gland inflammatory cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal B cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal B cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$sc_st=="Fetal Basophil_Mast")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Basophil_Mast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CD1C+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CD1C+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CEPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CEPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal CLEC9A+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal CLEC9A+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Collecting duct lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Collecting duct lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Connecting tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Connecting tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Distal tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Distal tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal EBMPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal EBMPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal EEPs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal EEPs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Enteroendocrine cells_intestine_pancreas")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Enteroendocrine cells_intestine_pancreas.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Enteroendocrine cells_stomach")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Enteroendocrine cells_stomach.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Erythroblast")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Erythroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal ETDs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal ETDs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal HSCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal HSCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal HSPCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal HSPCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal IL1B+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal IL1B+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal ILC 3")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal ILC 3.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Islet beta cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Islet beta cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Islet delta cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Islet delta cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal LEC")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal LEC.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Loop of Henle lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Loop of Henle lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Meg progenitors")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Meg progenitors.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Meg")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Meg.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Megakaryoblasts")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Megakaryoblasts.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Nephron progenitor cell")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Nephron progenitor cell")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal NK cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal NK cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal pDCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal pDCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Perivascular macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Perivascular macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Phagocytic macrophages")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Phagocytic macrophages.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Plasma cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Plasma cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Podocyte lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Podocyte lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Proximal tubule lineage")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Proximal tubule lineage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal PTPRC+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal PTPRC+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Pulmonary neuroendocrine cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Pulmonary neuroendocrine cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Renal vesicle")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Renal vesicle.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$sc_st=="Fetal S100A9+ DCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal S100A9+ DCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal T cells")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal T cells.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal TMEM119+ Microglia")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal TMEM119+ Microglia.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal TRAF1+ APCs")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal TRAF1+ APCs.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal Ureteric tip")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal Ureteric tip.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal adrenal")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal adrenal.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal brain")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal brain.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal kidney")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal kidney.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal liver")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal liver.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal lung")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal lung.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal placenta")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal placenta.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$sc_st=="Fetal spleen")
    
  {
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Fetal spleen.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult antigen presenting cell (RPS high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult antigen presenting cell (RPS high).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult astrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult astrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult AT2_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult AT2 cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult b_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult b_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult b_cell_plasmocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult b_cell_plasmocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult basal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult basal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult CB_CD34+"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult CB CD34+23.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult chondrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult chondrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult dendritic_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult dendritic_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial cell (APC)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial cell (APC).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial cell (endothelial to mesenchymal transition)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial cell (endothelial to mesenchymal transition).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult endothelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult endothelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult enterocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult enterocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult enterocyte_progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult enterocyte_progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult epithelial cell (intermediated)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult epithelial cell (intermediated).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult epithelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult epithelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult erythroid cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult erythroid cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult erythroid progenitor cell (RP high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult erythroid progenitor cell (RP high).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult fasciculata_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult fasciculata_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal acinar cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal acinar cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal chondrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal chondrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Fetal endocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal endocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal enterocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal enterocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Fetal epithelial progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal epithelial progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal fibroblast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal fibroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal skeletal muscle cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal skeletal muscle cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_mesenchymal_progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_mesenchymal_progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_neuron "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_neuron .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Fetal_stromal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Fetal_stromal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult Stromal_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult stromal_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult fibroblast"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult fibroblast.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult gastric chief cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult gastric chief cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult gastric endocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult gastric endocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult goblet_cell "){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult goblet_cell .tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult hepatocyte_Endodermal cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult hepatocyte_Endodermal cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult hESC87"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult hESC87.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult immature sertoli cell (Pre-Sertoli cell)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult immature sertoli cell (Pre-Sertoli cell).tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult intercalated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult intercalated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult intermediated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult intermediated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult kidney intercalated cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult kidney intercalated cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult loop of Henle"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult loop of Henle.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult M2 Macrophage"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult M2 Macrophage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult macrophage"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult macrophage.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult mast cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult mast cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult mesothelial cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult mesothelial cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult monocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Monocyte.csv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult myeloid cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult myeloid cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Neutrophil (RPS high)"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Neutrophil (RPS high)40.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult neutrophil"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.csv("Adult neutrophil.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult oligodendrocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult oligodendrocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult pancreas exocrine cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult pancreas exocrine cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult primordial germ cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult primordial germ cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult proliferating T cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult proliferating T cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult proximal tubule progenitor"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult proximal tubule progenitor.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult sinusoidal endothelial cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult sinusoidal endothelial cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult smooth_muscle_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult smooth_muscle_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult stratified_epithelial_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult stratified_epithelial_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  else if(input$st_sc =="Adult T_cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult T_cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult thyroid follicular cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult thyroid follicular cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  
  else if(input$st_sc =="Adult ureteric bud cell"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("dult ureteric bud cell.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  else if(input$st_sc =="Adult ventricle cardiomyocyte"){
    
    matric<-pb@assays$RNA@counts
    
    cells_sum <- Matrix::colSums(matric>=3)
    
    matric<-matric[,intersect(which(cells_sum>=stats::quantile(cells_sum,probs = 0.001)), which(cells_sum<=stats::quantile(cells_sum,probs = 1)))]
    
    matric<-matric[which(Matrix::rowSums(matric > 2) > 3),]
    
    cells_sum<-Matrix::rowSums(t(matric))
    
    matric<-Matrix::t(t(matric)/(cells_sum/stats::median(cells_sum)))
    r_name<-rownames(matric)
    media_genes <- read.table("Adult ventricle cardiomyocyte.tsv")
    same_genes<-intersect(r_name,media_genes[,1])
    if(length(same_genes)==1)
      message=paste("There is only one common gene ",same_genes," so connot compute stouffer score ",sep="")
    if(length(same_genes)==0)
      message="no common gene "
    if(length(same_genes)>1){
      matric= matric[same_genes,]####### didnot work why?
      
      #Further insuring have cells with atleast 10% expressed genes
      matric<-matric[,Matrix::colSums(matric>0)>(dim(matric)[1]/10)]
      print(dim(matric))
      seurat_RNA_matric<-matric
      print(sum(apply(seurat_RNA_matric,1,function(x) sum(x)==0)))
      print(sum(apply(seurat_RNA_matric,2,function(x) sum(x)==0)))
      print(dim(seurat_RNA_matric))
      ### First adding 1 (adding 1 to only zero, not only for zeros) then log2 then zscore
      seurat_RNA_matric[seurat_RNA_matric==0]=1
      seurat_RNA_matric<-log2(seurat_RNA_matric)
      seurat_RNA_matric<-t(apply(seurat_RNA_matric,1,function(x) x*mean(x)/sd(x)))
      ###Stouffer score and then boxplot for each cell type
      stouffer_score<- apply(seurat_RNA_matric,2,function(x) sum(x)/sqrt(length(x)))
      print(sum(is.na(stouffer_score)))
      cell_types<-Idents(pb)[names(stouffer_score)]
      unique_cell_types=unique(cell_types)
      Stouffer_score_df<-data.frame("Stouffer_score"=stouffer_score, "cell_types"=cell_types) 
      ggplot(Stouffer_score_df, aes(x=cell_types, y=Stouffer_score, fill=cell_types)) +
        geom_boxplot(position=position_dodge(.2)) +
        geom_jitter(shape=16, position=position_jitter(.1)) +
        theme_classic()+ theme(axis.text.x = element_text(angle = 90, hjust = 1))}}
  
  p_val_stouffer_score=matrix(0,nrow=length(unique_cell_types),ncol=length(unique_cell_types))
  colnames(p_val_stouffer_score)=unique_cell_types
  rownames(p_val_stouffer_score)=unique_cell_types
  for (i in 1:dim(p_val_stouffer_score)[1])
  {
    for (j in 1:dim(p_val_stouffer_score)[2])
    {
      print(paste(i,"_",j,sep=""))
      a=stouffer_score[names(cell_types)[cell_types==unique_cell_types[i]]]
      b=stouffer_score[names(cell_types)[cell_types==unique_cell_types[j]]]
      p_val_stouffer_score[i,j]=wilcox.test(a,b,alternative = "greater")$p.value
    }
  }
  
  format(round(p_val_stouffer_score, 4))
})

output$download_gtex_gt <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv(gtex_d(), file) 
    
  }) 
output$download_cell_gt <- downloadHandler(
  filename = function() {
    paste("data-", Sys.Date(), ".csv", sep="")
  },
  content = function(file) {
    write.csv(cell_grn_dt(), file) 
    
  }) 


output$cell_grnplot <- downloadHandler(
  filename = function(){'cell enrich.pdf'},
  content = function(file){
    pdf(file)
    print(      
      
      cell_grn_d()       
    )       
    dev.off()
  })






output$gtexplot <- downloadHandler(
  filename = function(){'gtex.pdf'},
  content = function(file){
    pdf(file)
    print(      
      
      stouffer_gtex()        
    )       
    dev.off()
  })


output$hpaplot <- downloadHandler(
  filename = function(){'hpa.pdf'},
  content = function(file){
    pdf(file)
    print(      
      
      stouffer_hpa()        
    )       
    dev.off()
  })



})